// @inrupt/vocab-solid/GeneratedVocab/OIDC.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * The OpenID Connect vocabulary used by the Solid-OIDC authentication specification.
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://www.w3.org/ns/solid/oidc#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * The OpenID Connect vocabulary used by the Solid-OIDC authentication specification.
 */
const OIDC = {
  PREFIX: "oidc",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "oidc": "http://www.w3.org/ns/solid/oidc#" },
  NS: _NS,

  // *******************
  // All the Properties.
  // *******************

  /**
   * TODO
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
   */
  application_type: _NS("application_type"),

  /**
   * A name for this client that can be presented to an End-User
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
   */
  client_name: _NS("client_name"),

  /**
   * A URI for this client's home page
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
   */
  client_uri: _NS("client_uri"),

  /**
   * A collection of URIs where individuals or entities responsible for this client can be contacted.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
   */
  contacts: _NS("contacts"),

  /**
   * TODO
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
   */
  default_max_age: _NS("default_max_age"),

  /**
   * A collection of OAuth 2.0 Grant Types that the client will restrict itself to using.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
   */
  grant_types: _NS("grant_types"),

  /**
   * A URI for this client's logo
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
   */
  logo_uri: _NS("logo_uri"),

  /**
   * A URI for this client's policy document
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
   */
  policy_uri: _NS("policy_uri"),

  /**
   * A collection of registered URIs used by the client for redirection
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
   */
  redirect_uris: _NS("redirect_uris"),

  /**
   * TODO
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
   */
  require_auth_time: _NS("require_auth_time"),

  /**
   * A collection of OAuth 2.0 Response Types that the client will restrict itself to using.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
   */
  response_types: _NS("response_types"),

  /**
   * An OAuth2 scope for the requested access token.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
   */
  scope: _NS("scope"),

  /**
   * The requested client authentication method for the Token Endpoint
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
   */
  token_endpoint_auth_method: _NS("token_endpoint_auth_method"),

  /**
   * A URI for this client's terms of service document
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
   */
  tos_uri: _NS("tos_uri"),

}

export default OIDC;
