// @inrupt/vocab-solid/GeneratedVocab/SOLID.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * The Solid Terms vocabulary defines terms referenced in Solid specifications.
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://www.w3.org/ns/solid/terms#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * The Solid Terms vocabulary defines terms referenced in Solid specifications.
 */
const SOLID = {
  PREFIX: "solid",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "solid": "http://www.w3.org/ns/solid/terms#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * A Solid account.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  Account: _NS("Account"),

  /**
   * A resource containing notifications.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  Inbox: _NS("Inbox"),

  /**
   * Listed Type Index is a registry of resources that are publicly discoverable by outside users and applications.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  ListedDocument: _NS("ListedDocument"),

  /**
   * A notification resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  Notification: _NS("Notification"),

  /**
   * A patch expresses conditional modifications to a resource that has an RDF-based representation.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  Patch: _NS("Patch"),

  /**
   * A resource containing time ordered items and sub-containers.  Sub-containers may be desirable in file based systems to split the timeline into logical components e.g. /yyyy-mm-dd/ as used in ISO 8061.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  Timeline: _NS("Timeline"),

  /**
   * A index of type registries for resources. Applications can register the RDF type they use and list them in the index resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  TypeIndex: _NS("TypeIndex"),

  /**
   * The registered types that map a RDF classes/types to their locations using either `instance` or `instanceContainer` property.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  TypeRegistration: _NS("TypeRegistration"),

  /**
   * Unlisted Type Index is a registry of resources that are private to the user and their apps, for types that are not publicly discoverable.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  UnlistedDocument: _NS("UnlistedDocument"),

  /**
   * The class of WebIDs - where a WebID is an HTTP URI which refers to an Agent (Person, Organization, Device, etc.).
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/solid#
   */
  WebId: _NS("WebId"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * A solid account belonging to an Agent.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  account: _NS("account"),

  /**
   * The triple patterns this patch removes from the document.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  deletes: _NS("deletes"),

  /**
   * A class that is used to map an listed or unlisted type index.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  forClass: _NS("forClass"),

  /**
   * Deprecated pointer to a Linked Data Notifications inbox; please use http://www.w3.org/ns/ldp#inbox instead.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  inbox: _NS("inbox"),

  /**
   * The triple patterns this patch adds to the document.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  inserts: _NS("inserts"),

  /**
   * Maps a type to an individual resource, typically an index or a directory listing resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  instance: _NS("instance"),

  /**
   * Maps a type to a container which the client would have to list to get the instances of that type.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  instanceContainer: _NS("instanceContainer"),

  /**
   * The login URI of a given server.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  loginEndpoint: _NS("loginEndpoint"),

  /**
   * The logout URI of a given server.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  logoutEndpoint: _NS("logoutEndpoint"),

  /**
   * Notification resource for an inbox.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  notification: _NS("notification"),

  /**
   * The preferred OpenID Connect issuer URI for a given WebID.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  oidcIssuer: _NS("oidcIssuer"),

  /**
   * The document to which this patch applies.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  patches: _NS("patches"),

  /**
   * Points to an unlisted type index resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  privateTypeIndex: _NS("privateTypeIndex"),

  /**
   * Points to a listed type index resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  publicTypeIndex: _NS("publicTypeIndex"),

  /**
   * Points to a TypeIndex resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  typeIndex: _NS("typeIndex"),

  /**
   * Indicates if a message has been read or not. This property should have a boolean datatype.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  read: _NS("read"),

  /**
   * The quota of non-volatile memory that is available for the account (in bytes)
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  storageQuota: _NS("storageQuota"),

  /**
   * The amount of non-volatile memory that the account have used (in bytes)
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  storageUsage: _NS("storageUsage"),

  /**
   * Timeline for a given resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  timeline: _NS("timeline"),

  /**
   * The conditions the document and the inserted and deleted triple patterns need to satisfy in order for the patch to be applied.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
   */
  where: _NS("where"),

  /**
   * A WebID is an HTTP URI which refers to an Agent (Person, Organization, Device, etc.).
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/solid#
   */
  webId: _NS("webId"),

  /**
   * A person or social entity that is considered to have control, rights, and responsibilities over a data storage.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/solid#
   */
  owner: _NS("owner"),

  /**
   * The IRI of the root resource of a Solid Pod.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/solid#
   */
  storageRoot: _NS("storageRoot"),

}

export default SOLID;
