// @inrupt/vocab-solid/GeneratedVocab/SOLID_SERVICE_CORE.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * The Solid vocabulary providing terms (e.g., classes, properties and text strings (e.g.
 informative labels or error messages) commonly used across Solid services.
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "https://inrupt.com/vocab/solid_service_core#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * The Solid vocabulary providing terms (e.g., classes, properties and text strings (e.g.
 informative labels or error messages) commonly used across Solid services.
 */
const SOLID_SERVICE_CORE = {
  PREFIX: "solid_service_core",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "solid_service_core": "https://inrupt.com/vocab/solid_service_core#" },
  NS: _NS,

  // *******************
  // All the Literals.
  // *******************

  /**
   * Failed to process incoming message as RDF
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: https://inrupt.com/vocab/solid_service_core#
   */
  errFailedToProcessIncomingRdf: _NS("errFailedToProcessIncomingRdf"),

  /**
   * Incoming request validated as RDF (it contains [{{0}}] triples!)
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: https://inrupt.com/vocab/solid_service_core#
   */
  msgRequestValidatedAsRdf: _NS("msgRequestValidatedAsRdf"),

}

export default SOLID_SERVICE_CORE;
