// @inrupt/vocab-solid/GeneratedVocab/SOLID_SD.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * The Solid Service Description vocabulary describes the services provided by a Solid Server.
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://www.w3.org/ns/solid/service-description#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * The Solid Service Description vocabulary describes the services provided by a Solid Server.
 */
const SOLID_SD = {
  PREFIX: "solid_sd",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "solid_sd": "http://www.w3.org/ns/solid/service-description#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * A Solid Service is some function provided by a Solid Server,
 e.g., to search a Pod, send a notification to an Inbox, logout of a Pod, etc.
   *
   * This term provides descriptions only in English.
   */
  Service: _NS("Service"),

  /**
   * An instance of solid_sd:Feature represents a feature of a
 Solid service. Specific types of features might include protocols (or
 sub-protocols)
   *
   * The term has a description only in English, with [1] label in the language [NoLocale], but [1] comment in the language [en] (so the difference is only between English and NoLocale, which we consider the same).
   */
  Feature: _NS("Feature"),

  /**
   * An instance of solid_sd:Protocol represents one specific
 protocol (or sub-protocol).
   *
   * The term has a description only in English, with [1] label in the language [NoLocale], but [1] comment in the language [en] (so the difference is only between English and NoLocale, which we consider the same).
   */
  Protocol: _NS("Protocol"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * A Solid Endpoint is the URL through which an agent requests
that a Solid Service performs its function, e.g., the SPARQL endpoint for a
"Search my Pod Solid Service.
   *
   * This term provides descriptions only in English.
   */
  endpoint: _NS("endpoint"),

  /**
   * Relates an instance of solid_sd:Service with a
 resource representing a supported feature.
   *
   * The term has a description only in English, with [1] label in the language [NoLocale], but [1] comment in the language [en] (so the difference is only between English and NoLocale, which we consider the same).
   */
  feature: _NS("feature"),

  /**
   * Relates an instance of solid_sd:Service to a protocol (or
 sub-protocol) that it implements (e.g., a WebSocket service might implement
 multiple, or just one, sub-protocols.
   *
   * This term provides descriptions only in English.
   */
  supportedProtocol: _NS("supportedProtocol"),

  /**
   * A Solid Service providing SPARQL query capability - could be
 across data within a single Pod, or across all Pods hosted by a Solid Server
 provider.
   *
   * This term provides descriptions only in English.
   */
  serviceSparql: _NS("serviceSparql"),

  /**
   * A Solid Service providing GraphQL query capability. This
 service initially provides authenticated and authorized access to sensitive
 access control data associated with resources in Pods.
   *
   * This term provides descriptions only in English.
   */
  serviceGraphql: _NS("serviceGraphql"),

  /**
   * A Solid Service providing the ability to 'power off' a Pod.
 This is useful in cases where a Pod owner wishes to completely disable all
 access to their entire Pod (e.g., in the case that they suspect access has been
 compromised).
   *
   * This term provides descriptions only in English.
   */
  servicePowerSwitch: _NS("servicePowerSwitch"),

  /**
   * A Solid Service providing the ability to register a new
 a Pod. Generally we’d expect this service to be provided by a Solid
 Identity Provider, although some server implementations may wish to offer this service directly themselves too.
   *
   * This term provides descriptions only in English.
   */
  servicePodRegistration: _NS("servicePodRegistration"),

  /**
   * A Solid Service providing the ability to subscribe to
 resource change events via a WebSocket connection.
   *
   * This term provides descriptions only in English.
   */
  serviceWebSocket: _NS("serviceWebSocket"),

  /**
   * solid-0.2
   *
   * This term has a label (in language [NoLocale]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * See also:
   *  - http://www.w3.org/ns/solid/service-description#WebSocketSubProtocol0_2
   */
  WebSocketSubProtocol0_2Label: _NS("WebSocketSubProtocol0_2Label"),

  /**
   * solid-0.1
   *
   * This term has a label (in language [NoLocale]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * See also:
   *  - http://www.w3.org/ns/solid/service-description#WebSocketSubProtocol0_1
   */
  WebSocketSubProtocol0_1Label: _NS("WebSocketSubProtocol0_1Label"),

}

export default SOLID_SD;
