// @inrupt/vocab-solid/GeneratedVocab/WS.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * This extension adds predicates making explicit some implicit knowledge in the Workspace ontology.
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://www.w3.org/ns/pim/space#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * This extension adds predicates making explicit some implicit knowledge in the Workspace ontology.
 */
const WS = {
  PREFIX: "ws",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "ws": "http://www.w3.org/ns/pim/space#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * A file describing a configuration. In this context, it is
 expected to describe the workspace configuration.
   *
   * This term provides multilingual descriptions, but has a mismatch between its labels and comments, with [2] labels in languages [en, fr], but [1] comment in the language [en].
   */
  ConfigurationFile: _NS("ConfigurationFile"),

  /**
   * A  storage is a space of URIs in which you can individually control for each resource
    who has access to it.

   *
   * This term provides descriptions only with no explicit locale.
   */
  ControlledStorage: _NS("ControlledStorage"),

  /**
   * A storage is a space of URIs in which you have access to data.

   *
   * This term provides descriptions only with no explicit locale.
   */
  Storage: _NS("Storage"),

  /**
   * This is a workspace for storing the
    information about the other workspaces.
    As a user, you normally don't have to worry about it.
   *
   * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
   */
  MasterWorkspace: _NS("MasterWorkspace"),

  /**
   * Access only by the you, the user.
   *
   * This term provides descriptions only with no explicit locale.
   */
  PrivateWorkspace: _NS("PrivateWorkspace"),

  /**
   * Workspaces are place where data is stored, and associated polices of privacy.
A given application typically stores information in several different
workspaces, some being user private, some shared, and some public.

   *
   * This term provides descriptions only with no explicit locale.
   */
  Workspace: _NS("Workspace"),

  /**
   * A personal storage is a space of URIs in which you and only you have access to data,
    you cannot give access to anyone else.

   *
   * This term provides descriptions only with no explicit locale.
   */
  PersonalStorage: _NS("PersonalStorage"),

  /**
   * Aceess may not be open to the public. Contains preferences resources.
   *
   * This term provides descriptions only with no explicit locale.
   */
  PreferencesWorkspace: _NS("PreferencesWorkspace"),

  /**
   * Access is to some but not all people.
   *
   * This term provides descriptions only with no explicit locale.
   */
  SharedWorkspace: _NS("SharedWorkspace"),

  /**
   * A public storage is a space of URIs in which you have access to data,
    and all data is accessible to anyone without control.

   *
   * This term provides descriptions only with no explicit locale.
   */
  PublicStorage: _NS("PublicStorage"),

  /**
   * Aceess is open to the public. Anything in a public workspace
    can be accesed by anyone.
   *
   * This term provides descriptions only with no explicit locale.
   */
  PublicWorkspace: _NS("PublicWorkspace"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * master workspace
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  masterWorkspace: _NS("masterWorkspace"),

  /**
   * preferences file
   *
   * This term has a label (in language [NoLocale]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  preferencesFile: _NS("preferencesFile"),

  /**
   * The storage in which this workspace is, or the storage which
contains this resource, or a storage available to this agent to use.
   *
   * This term provides descriptions only with no explicit locale.
   */
  storage: _NS("storage"),

  /**
   * URIs which start with this string are in this workspace or storage.
This may be used for constructing URIs for new storage resources.

   *
   * This term provides descriptions only with no explicit locale.
   */
  uriPrefix: _NS("uriPrefix"),

  /**
   * workspace
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  workspace: _NS("workspace"),

}

export default WS;
