// @inrupt/vocab-common-rdf/index.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */


export { default as ACL } from "./GeneratedVocab/ACL";

export { default as ALTR } from "./GeneratedVocab/ALTR";

export { default as AS } from "./GeneratedVocab/AS";

export { default as BOOKMARK } from "./GeneratedVocab/BOOKMARK";

export { default as CRED } from "./GeneratedVocab/CRED";

export { default as DCAT } from "./GeneratedVocab/DCAT";

export { default as DCTERMS } from "./GeneratedVocab/DCTERMS";

export { default as DOAP } from "./GeneratedVocab/DOAP";

export { default as EARL } from "./GeneratedVocab/EARL";

export { default as FOAF } from "./GeneratedVocab/FOAF";

export { default as HTTP } from "./GeneratedVocab/HTTP";

export { default as HTTPH_INRUPT } from "./GeneratedVocab/HTTPH_INRUPT";

export { default as HYDRA } from "./GeneratedVocab/HYDRA";

export { default as ICAL } from "./GeneratedVocab/ICAL";

export { default as JSONLD } from "./GeneratedVocab/JSONLD";

export { default as LDP } from "./GeneratedVocab/LDP";

export { default as ODRL } from "./GeneratedVocab/ODRL";

export { default as OLO } from "./GeneratedVocab/OLO";

export { default as OWL } from "./GeneratedVocab/OWL";

export { default as POSIX } from "./GeneratedVocab/POSIX";

export { default as PROV_O } from "./GeneratedVocab/PROV_O";

export { default as QB } from "./GeneratedVocab/QB";

export { default as RDF } from "./GeneratedVocab/RDF";

export { default as RDFS } from "./GeneratedVocab/RDFS";

export { default as RLOG } from "./GeneratedVocab/RLOG";

export { default as SCHEMA_INRUPT } from "./GeneratedVocab/SCHEMA_INRUPT";

export { default as SD } from "./GeneratedVocab/SD";

export { default as SDMX_DIMENSION } from "./GeneratedVocab/SDMX_DIMENSION";

export { default as SEC } from "./GeneratedVocab/SEC";

export { default as SHEX } from "./GeneratedVocab/SHEX";

export { default as SIOC } from "./GeneratedVocab/SIOC";

export { default as SKOS } from "./GeneratedVocab/SKOS";

export { default as SKOS_XL } from "./GeneratedVocab/SKOS_XL";

export { default as UI } from "./GeneratedVocab/UI";

export { default as UI_INRUPT } from "./GeneratedVocab/UI_INRUPT";

export { default as VANN } from "./GeneratedVocab/VANN";

export { default as VCARD } from "./GeneratedVocab/VCARD";

export { default as VOID } from "./GeneratedVocab/VOID";

export { default as VS } from "./GeneratedVocab/VS";

export { default as XSD } from "./GeneratedVocab/XSD";
