// @inrupt/vocab-solid/index.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */


export { default as ACL } from "./GeneratedVocab/ACL";

export { default as ACP } from "./GeneratedVocab/ACP";

export { default as OIDC } from "./GeneratedVocab/OIDC";

export { default as SOLID } from "./GeneratedVocab/SOLID";

export { default as SOLID_META_ACL } from "./GeneratedVocab/SOLID_META_ACL";

export { default as SOLID_META_CLIENT } from "./GeneratedVocab/SOLID_META_CLIENT";

export { default as SOLID_META_SERVER_CREATIONAL } from "./GeneratedVocab/SOLID_META_SERVER_CREATIONAL";

export { default as SOLID_META_SERVER_RUNTIME } from "./GeneratedVocab/SOLID_META_SERVER_RUNTIME";

export { default as SOLID_SD } from "./GeneratedVocab/SOLID_SD";

export { default as SOLID_SERVICE_CORE } from "./GeneratedVocab/SOLID_SERVICE_CORE";

export { default as WS } from "./GeneratedVocab/WS";
