// @inrupt/artifact-generator/test/Generated/UNIT_TEST/EndToEnd/module-version/Generated/SourceCodeArtifacts/JavaScript/GeneratedVocab/SCHEMA_INRUPT_EXT.js
/**
 * Generated by the Inrupt Artifact Generator [@inrupt/artifact-generator], version [3.1.1]
 * as part of artifact: [@inrupt/generated-vocab-schema-inrupt-ext], version: [1.0.5]
 * on 'Monday, October 2, 2023 11:19 AM'.
 *
 * Vocabulary built from input: [./test/resources/vocab/schema-snippet.ttl].
 * The generator detected the following terms in the source vocabulary:
 *  - Classes: [1]
 *  - Properties: [2]
 *  - Literals: [0]
 *  - Constant IRIs: [0]
 *  - Constant strings: [0]
 *
 * Inrupt extension to Schema.org terms.
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "https://schema.org/";
function _NS (localName) {
  return (_NAMESPACE + localName);
}

/**
 * Inrupt extension to Schema.org terms.
 *
 * Namespace IRI: [https://schema.org/]
 */
const SCHEMA_INRUPT_EXT = {
  PREFIX: "schema_inrupt_ext",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "schema_inrupt_ext": "https://schema.org/" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * A person (alive, dead, undead, or fictional).
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   */
  Person: _NS("Person"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * Given name. In the U.S., the first name of a Person. This can be used along with familyName instead of the name property.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   */
  givenName: _NS("givenName"),

  /**
   * Family name. In the U.S., the last name of an Person. This can be used along with givenName instead of the name property.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   */
  familyName: _NS("familyName"),

}

module.exports = SCHEMA_INRUPT_EXT;
