// @inrupt/artifact-generator/test/Generated/UNIT_TEST/EndToEnd/multiple-urls/Generated/SourceCodeArtifacts/JavaScript/GeneratedVocab/SCHEMA_INRUPT_EXT.js
/**
 * Generated by the Inrupt Artifact Generator [@inrupt/artifact-generator], version [3.1.1]
 * as part of artifact: [@inrupt/generated-vocab-schema-inrupt-ext], version: [1.0.0]
 * on 'Monday, October 2, 2023 11:19 AM'.
 *
 * Vocabulary built from inputs: [https://schema.org/Person.ttl, ./test/resources/vocab/schema-inrupt-ext.ttl].
 * The generator detected the following terms in the source vocabulary:
 *  - Classes: [3]
 *  - Properties: [73]
 *  - Literals: [0]
 *  - Constant IRIs: [0]
 *  - Constant strings: [0]
 *
 * Inrupt extension to Schema.org terms.
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "https://schema.org/";
function _NS (localName) {
  return (_NAMESPACE + localName);
}

/**
 * Inrupt extension to Schema.org terms.
 *
 * Namespace IRI: [https://schema.org/]
 */
const SCHEMA_INRUPT_EXT = {
  PREFIX: "schema_inrupt_ext",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "schema_inrupt_ext": "https://schema.org/" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * A person (alive, dead, undead, or fictional).
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   */
  Person: _NS("Person"),

  /**
   * The most generic type of item.
   *
   * This term provides descriptions only with no explicit locale.
   */
  Thing: _NS("Thing"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn't provide any 'rdfs:label', 'rdfs:comment', or 'dcterms:description', or 'skos:definition' metadata).
   */
  Patient: _NS("Patient"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * An alias for the item.
   *
   * This term has [6] labels and comments, in the languages [NoLocale, de, en, es, fr, it].
   */
  alternateName: _NS("alternateName"),

  /**
   * An additional name for a Person, can be used for a middle name.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   */
  additionalName: _NS("additionalName"),

  /**
   * An additional type for the item, typically used for adding more specific types from external vocabularies in microdata syntax. This is a relationship between something and a class that the thing is in. In RDFa syntax, it is better to use the native RDFa syntax - the 'typeof' attribute - for multiple types. Schema.org tools may have only weaker understanding of extra types, in particular those defined externally.
   *
   * This term provides descriptions only with no explicit locale.
   */
  additionalType: _NS("additionalType"),

  /**
   * Physical address of the item.
   *
   * This term provides descriptions only with no explicit locale.
   */
  address: _NS("address"),

  /**
   * An organization that this person is affiliated with. For example, a school/university, a club, or a team.
   *
   * This term provides descriptions only with no explicit locale.
   */
  affiliation: _NS("affiliation"),

  /**
   * An Organization (or ProgramMembership) to which this Person or Organization belongs.
   *
   * This term provides descriptions only with no explicit locale.
   */
  memberOf: _NS("memberOf"),

  /**
   * An organization that the person is an alumni of.
   *
   * This term provides descriptions only with no explicit locale.
   */
  alumniOf: _NS("alumniOf"),

  /**
   * Awards won by or for this item.
   *
   * This term provides descriptions only with no explicit locale.
   */
  awards: _NS("awards"),

  /**
   * An award won by or for this item.
   *
   * This term provides descriptions only with no explicit locale.
   */
  award: _NS("award"),

  /**
   * Date of birth.
   *
   * This term provides descriptions only with no explicit locale.
   */
  birthDate: _NS("birthDate"),

  /**
   * The place where the person was born.
   *
   * This term provides descriptions only with no explicit locale.
   */
  birthPlace: _NS("birthPlace"),

  /**
   * The brand(s) associated with a product or service, or the brand(s) maintained by an organization or business person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  brand: _NS("brand"),

  /**
   * A <a href="https://en.wikipedia.org/wiki/Call_sign">callsign</a>, as used in broadcasting and radio communications to identify people, radio and TV stations, or vehicles.
   *
   * This term provides descriptions only with no explicit locale.
   */
  callSign: _NS("callSign"),

  /**
   * The identifier property represents any kind of identifier for any kind of <a class="localLink" href="https://schema.org/Thing">Thing</a>, such as ISBNs, GTIN codes, UUIDs etc. Schema.org provides dedicated properties for representing many of these, either as textual strings or as URL (URI) links. See <a href="/docs/datamodel.html#identifierBg">background notes</a> for more details.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * See also:
   *  - https://schema.org/docs/datamodel.html#identifierBg
   */
  identifier: _NS("identifier"),

  /**
   * A child of the person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  children: _NS("children"),

  /**
   * A colleague of the person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  colleagues: _NS("colleagues"),

  /**
   * A colleague of the person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  colleague: _NS("colleague"),

  /**
   * A contact point for a person or organization.
   *
   * This term provides descriptions only with no explicit locale.
   */
  contactPoints: _NS("contactPoints"),

  /**
   * A contact point for a person or organization.
   *
   * This term provides descriptions only with no explicit locale.
   */
  contactPoint: _NS("contactPoint"),

  /**
   * Date of death.
   *
   * This term provides descriptions only with no explicit locale.
   */
  deathDate: _NS("deathDate"),

  /**
   * The place where the person died.
   *
   * This term provides descriptions only with no explicit locale.
   */
  deathPlace: _NS("deathPlace"),

  /**
   * A sub property of description. A short description of the item used to disambiguate from other, similar items. Information from other properties (in particular, name) may be necessary for the description to be useful for disambiguation.
   *
   * This term provides descriptions only with no explicit locale.
   */
  disambiguatingDescription: _NS("disambiguatingDescription"),

  /**
   * A description of the item.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   */
  description: _NS("description"),

  /**
   * The Dun &amp; Bradstreet DUNS number for identifying an organization or business person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  duns: _NS("duns"),

  /**
   * Email address.
   *
   * This term provides descriptions only with no explicit locale.
   */
  email: _NS("email"),

  /**
   * Family name. In the U.S., the last name of an Person. This can be used along with givenName instead of the name property.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   */
  familyName: _NS("familyName"),

  /**
   * The fax number.
   *
   * This term provides descriptions only with no explicit locale.
   */
  faxNumber: _NS("faxNumber"),

  /**
   * The most generic uni-directional social relation.
   *
   * This term provides descriptions only with no explicit locale.
   */
  follows: _NS("follows"),

  /**
   * A person or organization that supports (sponsors) something through some kind of financial contribution.
   *
   * This term provides descriptions only with no explicit locale.
   */
  funder: _NS("funder"),

  /**
   * A person or organization that supports a thing through a pledge, promise, or financial contribution. e.g. a sponsor of a Medical Study or a corporate sponsor of an event.
   *
   * This term provides descriptions only with no explicit locale.
   */
  sponsor: _NS("sponsor"),

  /**
   * Gender of something, typically a <a class="localLink" href="https://schema.org/Person">Person</a>, but possibly also fictional characters, animals, etc. While https://schema.org/Male and https://schema.org/Female may be used, text strings are also acceptable for people who do not identify as a binary gender. The <a class="localLink" href="https://schema.org/gender">gender</a> property can also be used in an extended sense to cover e.g. the gender of sports teams. As with the gender of individuals, we do not try to enumerate all possibilities. A mixed-gender <a class="localLink" href="https://schema.org/SportsTeam">SportsTeam</a> can be indicated with a text value of "Mixed".
   *
   * This term provides descriptions only with no explicit locale.
   */
  gender: _NS("gender"),

  /**
   * Given name. In the U.S., the first name of a Person. This can be used along with familyName instead of the name property.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   */
  givenName: _NS("givenName"),

  /**
   * The <a href="http://www.gs1.org/gln">Global Location Number</a> (GLN, sometimes also referred to as International Location Number or ILN) of the respective organization, person, or place. The GLN is a 13-digit number used to identify parties and physical locations.
   *
   * This term provides descriptions only with no explicit locale.
   */
  globalLocationNumber: _NS("globalLocationNumber"),

  /**
   * A credential awarded to the Person or Organization.
   *
   * This term provides descriptions only with no explicit locale.
   */
  hasCredential: _NS("hasCredential"),

  /**
   * The Person's occupation. For past professions, use Role for expressing dates.
   *
   * This term provides descriptions only with no explicit locale.
   */
  hasOccupation: _NS("hasOccupation"),

  /**
   * Indicates an OfferCatalog listing for this Organization, Person, or Service.
   *
   * This term provides descriptions only with no explicit locale.
   */
  hasOfferCatalog: _NS("hasOfferCatalog"),

  /**
   * Points-of-Sales operated by the organization or person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  hasPOS: _NS("hasPOS"),

  /**
   * The height of the item.
   *
   * This term provides descriptions only with no explicit locale.
   */
  height: _NS("height"),

  /**
   * A contact location for a person's residence.
   *
   * This term provides descriptions only with no explicit locale.
   */
  homeLocation: _NS("homeLocation"),

  /**
   * An honorific prefix preceding a Person's name such as Dr/Mrs/Mr.
   *
   * This term provides descriptions only with no explicit locale.
   */
  honorificPrefix: _NS("honorificPrefix"),

  /**
   * An honorific suffix preceding a Person's name such as M.D. /PhD/MSCSW.
   *
   * This term provides descriptions only with no explicit locale.
   */
  honorificSuffix: _NS("honorificSuffix"),

  /**
   * An image of the item. This can be a <a class="localLink" href="https://schema.org/URL">URL</a> or a fully described <a class="localLink" href="https://schema.org/ImageObject">ImageObject</a>.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * See also:
   *  - https://schema.org/URL
   *  - https://schema.org/ImageObject
   */
  image: _NS("image"),

  /**
   * The International Standard of Industrial Classification of All Economic Activities (ISIC), Revision 4 code for a particular organization, business person, or place.
   *
   * This term provides descriptions only with no explicit locale.
   */
  isicV4: _NS("isicV4"),

  /**
   * The job title of the person (for example, Financial Manager).
   *
   * This term provides descriptions only with no explicit locale.
   */
  jobTitle: _NS("jobTitle"),

  /**
   * The most generic bi-directional social/work relation.
   *
   * This term provides descriptions only with no explicit locale.
   */
  knows: _NS("knows"),

  /**
   * Of a <a class="localLink" href="https://schema.org/Person">Person</a>, and less typically of an <a class="localLink" href="https://schema.org/Organization">Organization</a>, to indicate a topic that is known about - suggesting possible expertise but not implying it. We do not distinguish skill levels here, or relate this to educational content, events, objectives or <a class="localLink" href="https://schema.org/JobPosting">JobPosting</a> descriptions.
   *
   * This term provides descriptions only with no explicit locale.
   */
  knowsAbout: _NS("knowsAbout"),

  /**
   * Of a <a class="localLink" href="https://schema.org/Person">Person</a>, and less typically of an <a class="localLink" href="https://schema.org/Organization">Organization</a>, to indicate a known language. We do not distinguish skill levels or reading/writing/speaking/signing here. Use language codes from the <a href="http://tools.ietf.org/html/bcp47">IETF BCP 47 standard</a>.
   *
   * This term provides descriptions only with no explicit locale.
   */
  knowsLanguage: _NS("knowsLanguage"),

  /**
   * Indicates a page (or other CreativeWork) for which this thing is the main entity being described. See <a href="/docs/datamodel.html#mainEntityBackground">background notes</a> for details.
   *
   * This term provides descriptions only with no explicit locale.
   */
  mainEntityOfPage: _NS("mainEntityOfPage"),

  /**
   * A pointer to products or services offered by the organization or person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  makesOffer: _NS("makesOffer"),

  /**
   * The North American Industry Classification System (NAICS) code for a particular organization or business person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  naics: _NS("naics"),

  /**
   * The name of the item.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   */
  name: _NS("name"),

  /**
   * Nationality of the person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  nationality: _NS("nationality"),

  /**
   * The total financial value of the person as calculated by subtracting assets from liabilities.
   *
   * This term provides descriptions only with no explicit locale.
   */
  netWorth: _NS("netWorth"),

  /**
   * Products owned by the organization or person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  owns: _NS("owns"),

  /**
   * A parents of the person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  parents: _NS("parents"),

  /**
   * A parent of this person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  parent: _NS("parent"),

  /**
   * Event that this person is a performer or participant in.
   *
   * This term provides descriptions only with no explicit locale.
   */
  performerIn: _NS("performerIn"),

  /**
   * Indicates a potential Action, which describes an idealized action in which this thing would play an 'object' role.
   *
   * This term provides descriptions only with no explicit locale.
   */
  potentialAction: _NS("potentialAction"),

  /**
   * The publishingPrinciples property indicates (typically via <a class="localLink" href="https://schema.org/URL">URL</a>) a document describing the editorial principles of an <a class="localLink" href="https://schema.org/Organization">Organization</a> (or individual e.g. a <a class="localLink" href="https://schema.org/Person">Person</a> writing a blog) that relate to their activities as a publisher, e.g. ethics or diversity policies. When applied to a <a class="localLink" href="https://schema.org/CreativeWork">CreativeWork</a> (e.g. <a class="localLink" href="https://schema.org/NewsArticle">NewsArticle</a>) the principles are those of the party primarily responsible for the creation of the <a class="localLink" href="https://schema.org/CreativeWork">CreativeWork</a>.<br/><br/>

While such policies are most typically expressed in natural language, sometimes related information (e.g. indicating a <a class="localLink" href="https://schema.org/funder">funder</a>) can be expressed using schema.org terminology.
   *
   * This term provides descriptions only with no explicit locale.
   */
  publishingPrinciples: _NS("publishingPrinciples"),

  /**
   * The most generic familial relation.
   *
   * This term provides descriptions only with no explicit locale.
   */
  relatedTo: _NS("relatedTo"),

  /**
   * URL of a reference Web page that unambiguously indicates the item's identity. E.g. the URL of the item's Wikipedia page, Wikidata entry, or official website.
   *
   * This term provides descriptions only with no explicit locale.
   */
  sameAs: _NS("sameAs"),

  /**
   * A pointer to products or services sought by the organization or person (demand).
   *
   * This term provides descriptions only with no explicit locale.
   */
  seeks: _NS("seeks"),

  /**
   * A sibling of the person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  siblings: _NS("siblings"),

  /**
   * A sibling of the person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  sibling: _NS("sibling"),

  /**
   * The person's spouse.
   *
   * This term provides descriptions only with no explicit locale.
   */
  spouse: _NS("spouse"),

  /**
   * A CreativeWork or Event about this Thing.
   *
   * This term provides descriptions only with no explicit locale.
   */
  subjectOf: _NS("subjectOf"),

  /**
   * The Tax / Fiscal ID of the organization or person, e.g. the TIN in the US or the CIF/NIF in Spain.
   *
   * This term provides descriptions only with no explicit locale.
   */
  taxID: _NS("taxID"),

  /**
   * The telephone number.
   *
   * This term provides descriptions only with no explicit locale.
   */
  telephone: _NS("telephone"),

  /**
   * URL of the item.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   */
  url: _NS("url"),

  /**
   * The Value-added Tax ID of the organization or person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  vatID: _NS("vatID"),

  /**
   * The weight of the product or person.
   *
   * This term provides descriptions only with no explicit locale.
   */
  weight: _NS("weight"),

  /**
   * A contact location for a person's place of work.
   *
   * This term provides descriptions only with no explicit locale.
   */
  workLocation: _NS("workLocation"),

  /**
   * Organizations that the person works for.
   *
   * This term provides descriptions only with no explicit locale.
   */
  worksFor: _NS("worksFor"),

}

module.exports = SCHEMA_INRUPT_EXT;
