// @inrupt/vocab-common-rdf/GeneratedVocab/ACL.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * [Generator provided] - Web Access Control Vocabulary
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://www.w3.org/ns/auth/acl#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * [Generator provided] - Web Access Control Vocabulary
 */
const ACL = {
  PREFIX: "acl",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "acl": "http://www.w3.org/ns/auth/acl#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * Any kind of access to a resource. Don't use this, use R W and RW
   *
   * The term has a description only in English, but has a mismatch between its labels and comments, with [0] labels, but [1] comment in the language [NoLocale].
   */
  Access: _NS("Access"),

  /**
   * Append accesses are specific write access which only add information, and do not remove information.
    For text files, for example, append access allows bytes to be added onto the end of the file.
    For RDF graphs, Append access allows adds triples to the graph but does not remove any.
    Append access is useful for dropbox functionality.
    Dropbox can be used for link notification, which the information added is a notification
    that a some link has been made elsewhere relevant to the given resource.
    
   *
   * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
   */
  Append: _NS("Append"),

  /**
   * write
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  Write: _NS("Write"),

  /**
   * A class of agents who have been authenticated.
In other words, anyone can access this resource, but not anonymously.
The social expectation is that the authentication process will provide an
identify and a name, or pseudonym.
(A new ID should not be minted for every access: the intent is that the user
is able to continue to use the ID for continues interactions with peers,
and for example to develop a reputation)

   *
   * This term provides descriptions only with no explicit locale.
   */
  AuthenticatedAgent: _NS("AuthenticatedAgent"),

  /**
   * An element of access control,
    allowing agent to agents access of some kind to resources or classes of resources
   *
   * This term provides descriptions only with no explicit locale.
   */
  Authorization: _NS("Authorization"),

  /**
   * Allows read/write access to the ACL for the resource(s)
   *
   * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
   */
  Control: _NS("Control"),

  /**
   * An Origin is basically a web site
        (Note WITHOUT the trailing slash after the domain name and port in its URI)
        and is the basis for controlling access to data by web apps
        in the Same Origin Model of web security.
        All scripts from the same origin are given the same right.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * See also:
   *  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin
   */
  Origin: _NS("Origin"),

  /**
   * The class of read operations
   *
   * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
   */
  Read: _NS("Read"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * The Access Control file for this information resource.
        This may of course be a virtual resource implemented by the access control system.
        Note also HTTP's header  Link:  foo.meta ;rel=meta can be used for this.
   *
   * This term provides descriptions only with no explicit locale.
   */
  accessControl: _NS("accessControl"),

  /**
   * The information resource to which access is being granted.
   *
   * This term provides descriptions only with no explicit locale.
   */
  accessTo: _NS("accessTo"),

  /**
   * A class of information resources to which access is being granted.
   *
   * This term provides descriptions only with no explicit locale.
   */
  accessToClass: _NS("accessToClass"),

  /**
   * A person or social entity to being given the right
   *
   * This term provides descriptions only with no explicit locale.
   */
  agent: _NS("agent"),

  /**
   * A class of persons or social entities to being given the right
   *
   * This term provides descriptions only with no explicit locale.
   */
  agentClass: _NS("agentClass"),

  /**
   * A group of persons or social entities to being given the right.
          The right is given to any entity which is a vcard:member of the group,
          as defined by the document received when the Group is dereferenced.
   *
   * This term provides descriptions only with no explicit locale.
   */
  agentGroup: _NS("agentGroup"),

  /**
   * If a resource has no ACL file (it is 404),
        then access to the resource if given by the ACL of the immediately
        containing directory, or failing that (404) the ACL of the recursively next
        containing directory which has an ACL file.
        Within that ACL file,
        any Authentication which has that directory as its acl:default applies to the
        resource. (The highest directory must have an ACL file.)

   *
   * This term provides descriptions only with no explicit locale.
   */
  default: _NS("default"),

  /**
   * THIS IS OBSOLETE AS OF 2017-08-01.   See 'default'.
        Was: A directory for which this authorization is used for new files in the directory.
   *
   * This term provides descriptions only with no explicit locale.
   */
  defaultForNew: _NS("defaultForNew"),

  /**
   * Delegates a person or another agent to act on behalf of the agent.
    For example, Alice delegates Bob to act on behalf of Alice for ACL purposes.
   *
   * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
   */
  delegates: _NS("delegates"),

  /**
   * A mode of access such as read or write.
   *
   * This term provides descriptions only with no explicit locale.
   */
  mode: _NS("mode"),

  /**
   * A web application, identified by its Origin, such as
        <https://scripts.example.com>, being given the right.
        When a user of the web application at a certain origin accesses the server,
        then the browser sets the Origin: header to warn that a possibly untrusted webapp
        is being used.
        Then, BOTH the user AND the origin must have the required access.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * See also:
   *  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin
   */
  origin: _NS("origin"),

  /**
   * The person or other agent which owns this.
    For example, the owner of a file in a filesystem.
    There is a sense of right to control.   Typically defaults to the agent who craeted
    something but can be changed.
   *
   * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
   */
  owner: _NS("owner"),

}

export default ACL;
