// @inrupt/vocab-common-rdf/GeneratedVocab/AS.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * Extended Activity Streams 2.0 Vocabulary
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "https://www.w3.org/ns/activitystreams#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * Extended Activity Streams 2.0 Vocabulary
 */
const AS = {
  PREFIX: "as",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "as": "https://www.w3.org/ns/activitystreams#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * An Object representing some form of Action that has been taken
   *
   * This term provides descriptions only in English.
   */
  Activity: _NS("Activity"),

  /**
   * Object
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  Object: _NS("Object"),

  /**
   * Represents a qualified reference to another resource. Patterned after the RFC5988 Web Linking Model
   *
   * This term provides descriptions only in English.
   */
  Link: _NS("Link"),

  /**
   * A rdf:List variant for Objects and Links
   *
   * This term provides descriptions only in English.
   */
  OrderedItems: _NS("OrderedItems"),

  /**
   * An ordered or unordered collection of Objects or Links
   *
   * This term provides descriptions only in English.
   */
  Collection: _NS("Collection"),

  /**
   * A subset of items from a Collection
   *
   * This term provides descriptions only in English.
   */
  CollectionPage: _NS("CollectionPage"),

  /**
   * An Image file
   *
   * This term provides descriptions only in English.
   */
  Image: _NS("Image"),

  /**
   * Represents a digital document/file of any sort
   *
   * This term provides descriptions only in English.
   */
  Document: _NS("Document"),

  /**
   * Represents a Social Graph relationship between two Individuals (indicated by the 'a' and 'b' properties)
   *
   * This term provides descriptions only in English.
   */
  Relationship: _NS("Relationship"),

  /**
   * A question of any sort.
   *
   * This term provides descriptions only in English.
   */
  Question: _NS("Question"),

  /**
   * An Activity that has no direct object
   *
   * This term provides descriptions only in English.
   */
  IntransitiveActivity: _NS("IntransitiveActivity"),

  /**
   * A Profile Document
   *
   * This term provides descriptions only in English.
   */
  Profile: _NS("Profile"),

  /**
   * A placeholder for a deleted object
   *
   * This term provides descriptions only in English.
   */
  Tombstone: _NS("Tombstone"),

  /**
   * A physical or logical location
   *
   * This term provides descriptions only in English.
   */
  Place: _NS("Place"),

  /**
   * An ordered subset of items from an OrderedCollection
   *
   * This term provides descriptions only in English.
   */
  OrderedCollectionPage: _NS("OrderedCollectionPage"),

  /**
   * A variation of Collection in which items are strictly ordered
   *
   * This term provides descriptions only in English.
   */
  OrderedCollection: _NS("OrderedCollection"),

  /**
   * Actor accepts the Object
   *
   * This term provides descriptions only in English.
   */
  Accept: _NS("Accept"),

  /**
   * Block
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  Block: _NS("Block"),

  /**
   * Actor is ignoring the Object
   *
   * This term provides descriptions only in English.
   */
  Ignore: _NS("Ignore"),

  /**
   * To Add an Object or Link to Something
   *
   * This term provides descriptions only in English.
   */
  Add: _NS("Add"),

  /**
   * Actor announces the object to the target
   *
   * This term provides descriptions only in English.
   */
  Announce: _NS("Announce"),

  /**
   * Represents a software application of any sort
   *
   * This term provides descriptions only in English.
   */
  Application: _NS("Application"),

  /**
   * To Arrive Somewhere (can be used, for instance, to indicate that a particular entity is currently located somewhere, e.g. a "check-in")
   *
   * This term provides descriptions only in English.
   */
  Arrive: _NS("Arrive"),

  /**
   * A written work. Typically several paragraphs long. For example, a blog post or a news article.
   *
   * This term provides descriptions only in English.
   */
  Article: _NS("Article"),

  /**
   * An audio file
   *
   * This term provides descriptions only in English.
   */
  Audio: _NS("Audio"),

  /**
   * To Create Something
   *
   * This term provides descriptions only in English.
   */
  Create: _NS("Create"),

  /**
   * To Delete Something
   *
   * This term provides descriptions only in English.
   */
  Delete: _NS("Delete"),

  /**
   * The actor dislikes the object
   *
   * This term provides descriptions only in English.
   */
  Dislike: _NS("Dislike"),

  /**
   * An Event of any kind
   *
   * This term provides descriptions only in English.
   */
  Event: _NS("Event"),

  /**
   * To flag something (e.g. flag as inappropriate, flag as spam, etc)
   *
   * This term provides descriptions only in English.
   */
  Flag: _NS("Flag"),

  /**
   * To Express Interest in Something
   *
   * This term provides descriptions only in English.
   */
  Follow: _NS("Follow"),

  /**
   * A Group of any kind.
   *
   * This term provides descriptions only in English.
   */
  Group: _NS("Group"),

  /**
   * To invite someone or something to something
   *
   * This term provides descriptions only in English.
   */
  Invite: _NS("Invite"),

  /**
   * To Offer something to someone or something
   *
   * This term provides descriptions only in English.
   */
  Offer: _NS("Offer"),

  /**
   * To Join Something
   *
   * This term provides descriptions only in English.
   */
  Join: _NS("Join"),

  /**
   * To Leave Something
   *
   * This term provides descriptions only in English.
   */
  Leave: _NS("Leave"),

  /**
   * To Like Something
   *
   * This term provides descriptions only in English.
   */
  Like: _NS("Like"),

  /**
   * The actor viewed the object
   *
   * This term provides descriptions only in English.
   */
  View: _NS("View"),

  /**
   * The actor listened to the object
   *
   * This term provides descriptions only in English.
   */
  Listen: _NS("Listen"),

  /**
   * The actor read the object
   *
   * This term provides descriptions only in English.
   */
  Read: _NS("Read"),

  /**
   * The actor is moving the object. The target specifies where the object is moving to. The origin specifies where the object is moving from.
   *
   * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
   */
  Move: _NS("Move"),

  /**
   * The actor is traveling to the target. The origin specifies where the actor is traveling from.
   *
   * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
   */
  Travel: _NS("Travel"),

  /**
   * A specialized Link that represents an @mention
   *
   * This term provides descriptions only in English.
   */
  Mention: _NS("Mention"),

  /**
   * A Short note, typically less than a single paragraph. A "tweet" is an example, or a "status update"
   *
   * This term provides descriptions only in English.
   */
  Note: _NS("Note"),

  /**
   * A Web Page
   *
   * This term provides descriptions only in English.
   */
  Page: _NS("Page"),

  /**
   * A Person
   *
   * This term provides descriptions only in English.
   */
  Person: _NS("Person"),

  /**
   * An Organization
   *
   * This term provides descriptions only in English.
   */
  Organization: _NS("Organization"),

  /**
   * Actor rejects the Object
   *
   * This term provides descriptions only in English.
   */
  Reject: _NS("Reject"),

  /**
   * To Remove Something
   *
   * This term provides descriptions only in English.
   */
  Remove: _NS("Remove"),

  /**
   * A service provided by some entity
   *
   * This term provides descriptions only in English.
   */
  Service: _NS("Service"),

  /**
   * Actor tentatively accepts the Object
   *
   * This term provides descriptions only in English.
   */
  TentativeAccept: _NS("TentativeAccept"),

  /**
   * Actor tentatively rejects the object
   *
   * This term provides descriptions only in English.
   */
  TentativeReject: _NS("TentativeReject"),

  /**
   * To Undo Something. This would typically be used to indicate that a previous Activity has been undone.
   *
   * This term provides descriptions only in English.
   */
  Undo: _NS("Undo"),

  /**
   * To Update/Modify Something
   *
   * This term provides descriptions only in English.
   */
  Update: _NS("Update"),

  /**
   * A Video document of any kind.
   *
   * This term provides descriptions only in English.
   */
  Video: _NS("Video"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * Subproperty of as:attributedTo that identifies the primary actor
   *
   * This term provides descriptions only in English.
   */
  actor: _NS("actor"),

  /**
   * Identifies an entity to which an object is attributed
   *
   * This term provides descriptions only in English.
   */
  attributedTo: _NS("attributedTo"),

  /**
   * attachment
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  attachment: _NS("attachment"),

  /**
   * attachments
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  attachments: _NS("attachments"),

  /**
   * Identifies the author of an object. Deprecated. Use as:attributedTo instead
   *
   * This term provides descriptions only in English.
   */
  author: _NS("author"),

  /**
   * bcc
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  bcc: _NS("bcc"),

  /**
   * bto
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  bto: _NS("bto"),

  /**
   * cc
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  cc: _NS("cc"),

  /**
   * Specifies the context within which an object exists or an activity was performed
   *
   * This term provides descriptions only in English.
   */
  context: _NS("context"),

  /**
   * current
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  current: _NS("current"),

  /**
   * first
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  first: _NS("first"),

  /**
   * generator
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  generator: _NS("generator"),

  /**
   * icon
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  icon: _NS("icon"),

  /**
   * image
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  image: _NS("image"),

  /**
   * inReplyTo
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  inReplyTo: _NS("inReplyTo"),

  /**
   * items
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  items: _NS("items"),

  /**
   * last
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  last: _NS("last"),

  /**
   * location
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  location: _NS("location"),

  /**
   * next
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  next: _NS("next"),

  /**
   * object
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  object: _NS("object"),

  /**
   * Describes a possible exclusive answer or option for a question.
   *
   * This term provides descriptions only in English.
   */
  oneOf: _NS("oneOf"),

  /**
   * Describes a possible inclusive answer or option for a question.
   *
   * This term provides descriptions only in English.
   */
  anyOf: _NS("anyOf"),

  /**
   * prev
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  prev: _NS("prev"),

  /**
   * preview
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  preview: _NS("preview"),

  /**
   * provider
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  provider: _NS("provider"),

  /**
   * replies
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  replies: _NS("replies"),

  /**
   * result
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  result: _NS("result"),

  /**
   * audience
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  audience: _NS("audience"),

  /**
   * partOf
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  partOf: _NS("partOf"),

  /**
   * tag
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  tag: _NS("tag"),

  /**
   * tags
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  tags: _NS("tags"),

  /**
   * target
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  target: _NS("target"),

  /**
   * For certain activities, specifies the entity from which the action is directed.
   *
   * This term provides descriptions only in English.
   */
  origin: _NS("origin"),

  /**
   * Indentifies an object used (or to be used) to complete an activity
   *
   * This term provides descriptions only in English.
   */
  instrument: _NS("instrument"),

  /**
   * to
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  to: _NS("to"),

  /**
   * Specifies a link to a specific representation of the Object
   *
   * This term provides descriptions only in English.
   */
  url: _NS("url"),

  /**
   * On a Relationship object, identifies the subject. e.g. when saying "John is connected to Sally", 'subject' refers to 'John'
   *
   * This term provides descriptions only in English.
   */
  subject: _NS("subject"),

  /**
   * On a Relationship object, describes the type of relationship
   *
   * This term provides descriptions only in English.
   */
  relationship: _NS("relationship"),

  /**
   * On a Profile object, describes the object described by the profile
   *
   * This term provides descriptions only in English.
   */
  describes: _NS("describes"),

  /**
   * On a Tombstone object, describes the former type of the deleted object
   *
   * This term provides descriptions only in English.
   */
  formerType: _NS("formerType"),

  /**
   * Specifies the accuracy around the point established by the longitude and latitude
   *
   * This term provides descriptions only in English.
   */
  accuracy: _NS("accuracy"),

  /**
   * The altitude of a place
   *
   * This term provides descriptions only in English.
   */
  altitude: _NS("altitude"),

  /**
   * The content of the object.
   *
   * This term provides descriptions only in English.
   */
  content: _NS("content"),

  /**
   * name
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  name: _NS("name"),

  /**
   * downstreamDuplicates
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  downstreamDuplicates: _NS("downstreamDuplicates"),

  /**
   * The duration of the object
   *
   * This term provides descriptions only in English.
   */
  duration: _NS("duration"),

  /**
   * The ending time of the object
   *
   * This term provides descriptions only in English.
   */
  endTime: _NS("endTime"),

  /**
   * The display height expressed as device independent pixels
   *
   * This term provides descriptions only in English.
   */
  height: _NS("height"),

  /**
   * The target URI of the Link
   *
   * This term provides descriptions only in English.
   */
  href: _NS("href"),

  /**
   * A hint about the language of the referenced resource
   *
   * This term provides descriptions only in English.
   */
  hreflang: _NS("hreflang"),

  /**
   * id
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  id: _NS("id"),

  /**
   * The latitude
   *
   * This term provides descriptions only in English.
   */
  latitude: _NS("latitude"),

  /**
   * The longitude
   *
   * This term provides descriptions only in English.
   */
  longitude: _NS("longitude"),

  /**
   * The MIME Media Type
   *
   * This term provides descriptions only in English.
   */
  mediaType: _NS("mediaType"),

  /**
   * objectType
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  objectType: _NS("objectType"),

  /**
   * Specifies the date and time the object was published
   *
   * This term provides descriptions only in English.
   */
  published: _NS("published"),

  /**
   * Specifies a radius around the point established by the longitude and latitude
   *
   * This term provides descriptions only in English.
   */
  radius: _NS("radius"),

  /**
   * A numeric rating (>= 0.0, <= 5.0) for the object
   *
   * This term provides descriptions only in English.
   */
  rating: _NS("rating"),

  /**
   * The RFC 5988 or HTML5 Link Relation associated with the Link
   *
   * This term provides descriptions only in English.
   */
  rel: _NS("rel"),

  /**
   * In a strictly ordered logical collection, specifies the index position of the first item in the items list
   *
   * This term provides descriptions only in English.
   */
  startIndex: _NS("startIndex"),

  /**
   * The starting time of the object
   *
   * This term provides descriptions only in English.
   */
  startTime: _NS("startTime"),

  /**
   * A short summary of the object
   *
   * This term provides descriptions only in English.
   */
  summary: _NS("summary"),

  /**
   * The total number of items in a logical collection
   *
   * This term provides descriptions only in English.
   */
  totalItems: _NS("totalItems"),

  /**
   * Identifies the unit of measurement used by the radius, altitude and accuracy properties. The value can be expressed either as one of a set of predefined units or as a well-known common URI that identifies units.
   *
   * This term provides descriptions only in English.
   */
  units: _NS("units"),

  /**
   * Specifies when the object was last updated
   *
   * This term provides descriptions only in English.
   */
  updated: _NS("updated"),

  /**
   * upstreamDuplicates
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  upstreamDuplicates: _NS("upstreamDuplicates"),

  /**
   * verb
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  verb: _NS("verb"),

  /**
   * Specifies the preferred display width of the content, expressed in terms of device independent pixels.
   *
   * This term provides descriptions only in English.
   */
  width: _NS("width"),

  /**
   * Specifies the date and time the object was deleted
   *
   * This term provides descriptions only in English.
   */
  deleted: _NS("deleted"),

}

export default AS;
