// @inrupt/vocab-common-rdf/GeneratedVocab/BOOKMARK.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * [Generator provided] - W3C Bookmark vocabulary
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://www.w3.org/2002/01/bookmark#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * [Generator provided] - W3C Bookmark vocabulary
 */
const BOOKMARK = {
  PREFIX: "bookmark",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "bookmark": "http://www.w3.org/2002/01/bookmark#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * Topic
   *
   * This term has a label (in language [NoLocale]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  Topic: _NS("Topic"),

  /**
   * Specifies a behavior; when the
                object of type 'Shortcut' is activated, the
                client follows the 'recalls' property and
                activates the object at the end of that
                'recalls' property.  The target object may
                be another Bookmark or may be a Topic.
   *
   * This term provides descriptions only with no explicit locale.
   */
  Shortcut: _NS("Shortcut"),

  /**
   * The class to which all bookmarks belong.
   *
   * This term provides descriptions only with no explicit locale.
   */
  Bookmark: _NS("Bookmark"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * relates a bookmark to a topic.
                A bookmark must have at least one hasTopic property.
                The typical user operation of following a bookmark link
                will use the value of the b:recalls property. This property
                corresponds to XBEL:href property.
   *
   * This term provides descriptions only with no explicit locale.
   */
  hasTopic: _NS("hasTopic"),

  /**
   * connects a Shortcut to the bookmark or topic that is
being included by reference in some other topic
   *
   * The term has a description only in English, but has a mismatch between its labels and comments, with [0] labels, but [1] comment in the language [en].
   */
  leadsTo: _NS("leadsTo"),

  /**
   * This corresponds to XBEL:href an
                object of type Bookmark is expected to have a
                'recalls' relationship to the document being
                bookmarked.  The 'bookmarks' property is an older
                name for the 'recalls' relationship.
   *
   * This term provides descriptions only with no explicit locale.
   */
  bookmarks: _NS("bookmarks"),

  /**
   * Relates a bookmark with the resource that
                has been bookmarked.  This corresponds to XBEL:href;
                an object of type Bookmark is expected to have a
                'recalls' relationship to the document being
                bookmarked 
   *
   * This term provides descriptions only with no explicit locale.
   */
  recalls: _NS("recalls"),

  /**
   * Describes a relationship between Topics.
        When a topic T is a sub-topic of a topic U then all
        bookmarks that have topic T are also considered to have
        topic U. A topic may be a sub-topic of one or more
        topics; trivially, every topic is a sub-topic of itself.
        More formally; for all B, T, and U: b b:hasTopic T,
        T b:subTopicOf U implies B b:hasTopic U.
   *
   * This term provides descriptions only with no explicit locale.
   */
  subTopicOf: _NS("subTopicOf"),

}

export default BOOKMARK;
