// @inrupt/vocab-common-rdf/GeneratedVocab/CRED.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * A vocabulary for the Data Model for W3C Verifiable Credentials.
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "https://www.w3.org/2018/credentials#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * A vocabulary for the Data Model for W3C Verifiable Credentials.
 */
const CRED = {
  PREFIX: "cred",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "cred": "https://www.w3.org/2018/credentials#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  VerifiableCredential: _NS("VerifiableCredential"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  JsonSchemaValidator2018: _NS("JsonSchemaValidator2018"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  ManualRefreshService2018: _NS("ManualRefreshService2018"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  VerifiablePresentation: _NS("VerifiablePresentation"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  verifiableCredential: _NS("verifiableCredential"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  credentialSchema: _NS("credentialSchema"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  credentialStatus: _NS("credentialStatus"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  credentialSubject: _NS("credentialSubject"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  evidence: _NS("evidence"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  expirationDate: _NS("expirationDate"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  holder: _NS("holder"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  issued: _NS("issued"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  issuer: _NS("issuer"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  issuanceDate: _NS("issuanceDate"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  refreshService: _NS("refreshService"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  termsOfUse: _NS("termsOfUse"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  validFrom: _NS("validFrom"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://www.w3.org/2018/credentials#
   */
  validUntil: _NS("validUntil"),

}

export default CRED;
