// @inrupt/vocab-common-rdf/GeneratedVocab/DCTERMS.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * [Generator provided] - Dublin Core Terms - for describing resources
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://purl.org/dc/terms/";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * [Generator provided] - Dublin Core Terms - for describing resources
 */
const DCTERMS = {
  PREFIX: "dcterms",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "dcterms": "http://purl.org/dc/terms/" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * A resource that acts or has the power to act.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  Agent: _NS("Agent"),

  /**
   * A group of agents.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  AgentClass: _NS("AgentClass"),

  /**
   * A book, article, or other documentary resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  BibliographicResource: _NS("BibliographicResource"),

  /**
   * A digital resource format.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  FileFormat: _NS("FileFormat"),

  /**
   * A file format or physical medium.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  MediaType: _NS("MediaType"),

  /**
   * A media type or extent.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  MediaTypeOrExtent: _NS("MediaTypeOrExtent"),

  /**
   * A rate at which something recurs.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  Frequency: _NS("Frequency"),

  /**
   * The extent or range of judicial, law enforcement, or other authority.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  Jurisdiction: _NS("Jurisdiction"),

  /**
   * A location, period of time, or jurisdiction.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  LocationPeriodOrJurisdiction: _NS("LocationPeriodOrJurisdiction"),

  /**
   * A legal document giving official permission to do something with a Resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  LicenseDocument: _NS("LicenseDocument"),

  /**
   * A statement about the intellectual property rights (IPR) held in or over a Resource, a legal document giving official permission to do something with a resource, or a statement about access rights.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  RightsStatement: _NS("RightsStatement"),

  /**
   * A system of signs, symbols, sounds, gestures, or rules used in communication.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  LinguisticSystem: _NS("LinguisticSystem"),

  /**
   * A spatial region or named place.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  Location: _NS("Location"),

  /**
   * A method by which resources are added to a collection.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  MethodOfAccrual: _NS("MethodOfAccrual"),

  /**
   * A process that is used to engender knowledge, attitudes, and skills.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  MethodOfInstruction: _NS("MethodOfInstruction"),

  /**
   * An interval of time that is named or defined by its start and end dates.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  PeriodOfTime: _NS("PeriodOfTime"),

  /**
   * A physical material or carrier.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  PhysicalMedium: _NS("PhysicalMedium"),

  /**
   * A material thing.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  PhysicalResource: _NS("PhysicalResource"),

  /**
   * A plan or course of action by an authority, intended to influence and determine decisions, actions, and other matters.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  Policy: _NS("Policy"),

  /**
   * A statement of any changes in ownership and custody of a resource since its creation that are significant for its authenticity, integrity, and interpretation.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  ProvenanceStatement: _NS("ProvenanceStatement"),

  /**
   * A dimension or extent, or a time taken to play or execute.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  SizeOrDuration: _NS("SizeOrDuration"),

  /**
   * A basis for comparison; a reference point against which other things can be evaluated.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  Standard: _NS("Standard"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * Date on which the resource was changed.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  modified: _NS("modified"),

  /**
   * An entity responsible for making the resource available.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  publisher: _NS("publisher"),

  /**
   * A name given to the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  title: _NS("title"),

  /**
   * A related resource that is a version, edition, or adaptation of the described resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  hasVersion: _NS("hasVersion"),

  /**
   * Date of formal issuance (e.g., publication) of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  issued: _NS("issued"),

  /**
   * A point or period of time associated with an event in the lifecycle of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  date: _NS("date"),

  /**
   * An account of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  description: _NS("description"),

  /**
   * A related resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  relation: _NS("relation"),

  /**
   * The set of regions in space defined by their geographic coordinates according to the DCMI Box Encoding Scheme.
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://dublincore.org/documents/dcmi-box/
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  Box: _NS("Box"),

  /**
   * The set of codes listed in ISO 3166-1 for the representation of names of countries.
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://www.iso.org/iso/en/prods-services/iso3166ma/02iso-3166-code-lists/list-en1.html
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  ISO3166: _NS("ISO3166"),

  /**
   * The three-letter alphabetic codes listed in ISO639-2 for the representation of names of languages.
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://lcweb.loc.gov/standards/iso639-2/langhome.html
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  ISO639_2: _NS("ISO639-2"),

  /**
   * The set of three-letter codes listed in ISO 639-3 for the representation of names of languages.
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://www.sil.org/iso639-3/
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  ISO639_3: _NS("ISO639-3"),

  /**
   * The set of time intervals defined by their limits according to the DCMI Period Encoding Scheme.
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://dublincore.org/documents/dcmi-period/
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  Period: _NS("Period"),

  /**
   * The set of points in space defined by their geographic coordinates according to the DCMI Point Encoding Scheme.
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://dublincore.org/documents/dcmi-point/
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  Point: _NS("Point"),

  /**
   * The set of tags, constructed according to RFC 1766, for the identification of languages.
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://www.ietf.org/rfc/rfc1766.txt
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  RFC1766: _NS("RFC1766"),

  /**
   * The set of tags constructed according to RFC 3066 for the identification of languages.
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://www.ietf.org/rfc/rfc3066.txt
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  RFC3066: _NS("RFC3066"),

  /**
   * The set of tags constructed according to RFC 4646 for the identification of languages.
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://www.ietf.org/rfc/rfc4646.txt
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  RFC4646: _NS("RFC4646"),

  /**
   * The set of tags constructed according to RFC 5646 for the identification of languages.
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://www.ietf.org/rfc/rfc5646.txt
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  RFC5646: _NS("RFC5646"),

  /**
   * The set of identifiers constructed according to the generic syntax for Uniform Resource Identifiers as specified by the Internet Engineering Task Force.
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://www.ietf.org/rfc/rfc3986.txt
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  URI: _NS("URI"),

  /**
   * The set of dates and times constructed according to the W3C Date and Time Formats Specification.
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://www.w3.org/TR/NOTE-datetime
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  W3CDTF: _NS("W3CDTF"),

  /**
   * A summary of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  abstract: _NS("abstract"),

  /**
   * Information about who can access the resource or an indication of its security status.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  accessRights: _NS("accessRights"),

  /**
   * Information about rights held in and over the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  rights: _NS("rights"),

  /**
   * The method by which items are added to a collection.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  accrualMethod: _NS("accrualMethod"),

  /**
   * The frequency with which items are added to a collection.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  accrualPeriodicity: _NS("accrualPeriodicity"),

  /**
   * The policy governing the addition of items to a collection.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  accrualPolicy: _NS("accrualPolicy"),

  /**
   * An alternative name for the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  alternative: _NS("alternative"),

  /**
   * A class of entity for whom the resource is intended or useful.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  audience: _NS("audience"),

  /**
   * Date (often a range) that the resource became or will become available.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  available: _NS("available"),

  /**
   * A bibliographic reference for the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  bibliographicCitation: _NS("bibliographicCitation"),

  /**
   * An unambiguous reference to the resource within a given context.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  identifier: _NS("identifier"),

  /**
   * An established standard to which the described resource conforms.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  conformsTo: _NS("conformsTo"),

  /**
   * An entity responsible for making contributions to the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  contributor: _NS("contributor"),

  /**
   * The spatial or temporal topic of the resource, the spatial applicability of the resource, or the jurisdiction under which the resource is relevant.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  coverage: _NS("coverage"),

  /**
   * Date of creation of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  created: _NS("created"),

  /**
   * An entity primarily responsible for making the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  creator: _NS("creator"),

  /**
   * Date of acceptance of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  dateAccepted: _NS("dateAccepted"),

  /**
   * Date of copyright.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  dateCopyrighted: _NS("dateCopyrighted"),

  /**
   * Date of submission of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  dateSubmitted: _NS("dateSubmitted"),

  /**
   * A class of entity, defined in terms of progression through an educational or training context, for which the described resource is intended.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  educationLevel: _NS("educationLevel"),

  /**
   * The size or duration of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  extent: _NS("extent"),

  /**
   * The file format, physical medium, or dimensions of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  format: _NS("format"),

  /**
   * A related resource that is substantially the same as the pre-existing described resource, but in another format.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  hasFormat: _NS("hasFormat"),

  /**
   * A related resource that is included either physically or logically in the described resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  hasPart: _NS("hasPart"),

  /**
   * A process, used to engender knowledge, attitudes and skills, that the described resource is designed to support.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  instructionalMethod: _NS("instructionalMethod"),

  /**
   * A related resource that is substantially the same as the described resource, but in another format.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  isFormatOf: _NS("isFormatOf"),

  /**
   * A related resource in which the described resource is physically or logically included.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  isPartOf: _NS("isPartOf"),

  /**
   * A related resource that references, cites, or otherwise points to the described resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  isReferencedBy: _NS("isReferencedBy"),

  /**
   * A related resource that supplants, displaces, or supersedes the described resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  isReplacedBy: _NS("isReplacedBy"),

  /**
   * A related resource that requires the described resource to support its function, delivery, or coherence.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  isRequiredBy: _NS("isRequiredBy"),

  /**
   * A related resource of which the described resource is a version, edition, or adaptation.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  isVersionOf: _NS("isVersionOf"),

  /**
   * A language of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  language: _NS("language"),

  /**
   * A legal document giving official permission to do something with the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  license: _NS("license"),

  /**
   * An entity that mediates access to the resource and for whom the resource is intended or useful.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  mediator: _NS("mediator"),

  /**
   * The material or physical carrier of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  medium: _NS("medium"),

  /**
   * A statement of any changes in ownership and custody of the resource since its creation that are significant for its authenticity, integrity, and interpretation.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  provenance: _NS("provenance"),

  /**
   * A related resource that is referenced, cited, or otherwise pointed to by the described resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  references: _NS("references"),

  /**
   * A related resource that is supplanted, displaced, or superseded by the described resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  replaces: _NS("replaces"),

  /**
   * A related resource that is required by the described resource to support its function, delivery, or coherence.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  requires: _NS("requires"),

  /**
   * A person or organization owning or managing rights over the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  rightsHolder: _NS("rightsHolder"),

  /**
   * A related resource from which the described resource is derived.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  source: _NS("source"),

  /**
   * Spatial characteristics of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  spatial: _NS("spatial"),

  /**
   * The topic of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  subject: _NS("subject"),

  /**
   * A list of subunits of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  tableOfContents: _NS("tableOfContents"),

  /**
   * Temporal characteristics of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  temporal: _NS("temporal"),

  /**
   * The nature or genre of the resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  type: _NS("type"),

  /**
   * Date (often a range) of validity of a resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/dc/terms/
   */
  valid: _NS("valid"),

}

export default DCTERMS;
