// @inrupt/vocab-common-rdf/GeneratedVocab/DOAP.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * [Generator provided] - Description of a Project (DOAP) vocabulary
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://usefulinc.com/ns/doap#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * [Generator provided] - Description of a Project (DOAP) vocabulary
 */
const DOAP = {
  PREFIX: "doap",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "doap": "http://usefulinc.com/ns/doap#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * A project.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  Project: _NS("Project"),

  /**
   * Version information of a project release.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  Version: _NS("Version"),

  /**
   * A specification of a system's aspects, technical or otherwise.
   *
   * This term has [2] labels and comments, in the languages [en, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  Specification: _NS("Specification"),

  /**
   * Source code repository.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  Repository: _NS("Repository"),

  /**
   * Subversion source code repository.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  SVNRepository: _NS("SVNRepository"),

  /**
   * Git source code repository.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  GitRepository: _NS("GitRepository"),

  /**
   * BitKeeper source code repository.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  BKRepository: _NS("BKRepository"),

  /**
   * CVS source code repository.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  CVSRepository: _NS("CVSRepository"),

  /**
   * GNU Arch source code repository.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  ArchRepository: _NS("ArchRepository"),

  /**
   * Bazaar source code branch.
   *
   * This term has [2] labels and comments, in the languages [en, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  BazaarBranch: _NS("BazaarBranch"),

  /**
   * Git source code branch.
   *
   * This term has [2] labels and comments, in the languages [en, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  GitBranch: _NS("GitBranch"),

  /**
   * Mercurial source code repository.
   *
   * This term has [2] labels and comments, in the languages [en, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  HgRepository: _NS("HgRepository"),

  /**
   * darcs source code repository.
   *
   * This term has [4] labels and comments, in the languages [en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  DarcsRepository: _NS("DarcsRepository"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * A name of something.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  name: _NS("name"),

  /**
   * URL of a project's homepage,
		associated with exactly one project.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  homepage: _NS("homepage"),

  /**
   * URL of a project's past homepage,
		associated with exactly one project.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  old_homepage: _NS("old-homepage"),

  /**
   * Date when something was created, in YYYY-MM-DD form. e.g. 2004-04-05
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  created: _NS("created"),

  /**
   * Short (8 or 9 words) plain text description of a project.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  shortdesc: _NS("shortdesc"),

  /**
   * Plain text description of a project, of 2-4 sentences in length.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  description: _NS("description"),

  /**
   * A project release.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  release: _NS("release"),

  /**
   * Mailing list home page or email address.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  mailing_list: _NS("mailing-list"),

  /**
   * A forum or community that supports this project.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  support_forum: _NS("support-forum"),

  /**
   * A forum or community for developers of this project.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  developer_forum: _NS("developer-forum"),

  /**
   * A category of project.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  category: _NS("category"),

  /**
   * The URI of an RDF description of the license the software is distributed under. E.g. a SPDX reference
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  license: _NS("license"),

  /**
   * Source code repository.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  repository: _NS("repository"),

  /**
   * The project that uses a repository.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  repositoryOf: _NS("repositoryOf"),

  /**
   * Repository for anonymous access.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  anon_root: _NS("anon-root"),

  /**
   * Web browser interface to repository.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  browse: _NS("browse"),

  /**
   * Module name of a Subversion, CVS, BitKeeper or Arch repository.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  module: _NS("module"),

  /**
   * Location of a repository.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  location: _NS("location"),

  /**
   * Web page from which the project software can be downloaded.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  download_page: _NS("download-page"),

  /**
   * Mirror of software download web page.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  download_mirror: _NS("download-mirror"),

  /**
   * Revision identifier of a software release.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  revision: _NS("revision"),

  /**
   * URI of download associated with this release.
   *
   * This term has [3] labels and comments, in the languages [cs, en, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  file_release: _NS("file-release"),

  /**
   * URL of Wiki for collaborative discussion of project.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  wiki: _NS("wiki"),

  /**
   * Bug tracker for a project.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  bug_database: _NS("bug-database"),

  /**
   * Web page with screenshots of project.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  screenshots: _NS("screenshots"),

  /**
   * Maintainer of a project, a project leader.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  maintainer: _NS("maintainer"),

  /**
   * Developer of software for the project.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  developer: _NS("developer"),

  /**
   * Contributor of documentation to the project.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  documenter: _NS("documenter"),

  /**
   * Contributor of translations to the project.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  translator: _NS("translator"),

  /**
   * A tester or other quality control contributor.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  tester: _NS("tester"),

  /**
   * Project contributor.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  helper: _NS("helper"),

  /**
   * Programming language a project is implemented in or intended for use with.
   *
   * This term has [6] labels and comments, in the languages [cs, de, en, es, fr, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  programming_language: _NS("programming-language"),

  /**
   * Operating system that a project is limited to.  Omit this property if the project is not OS-specific.
   *
   * This term provides multilingual descriptions, but has a mismatch between its labels and comments, with [6] labels in languages [cs, de, en, es, fr, pt], but [5] comments in languages [cs, de, en, es, fr].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  os: _NS("os"),

  /**
   * A specification that a project implements. Could be a standard, API or legally defined level of conformance.
   *
   * This term has [2] labels and comments, in the languages [en, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  implements: _NS("implements"),

  /**
   * The URI of a web service endpoint where software as a service may be accessed
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  service_endpoint: _NS("service-endpoint"),

  /**
   * ISO language code a project has been translated into
   *
   * This term has [2] labels and comments, in the languages [en, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  language: _NS("language"),

  /**
   * Vendor organization: commercial, free or otherwise
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  vendor: _NS("vendor"),

  /**
   * Indicator of software platform (non-OS specific), e.g. Java, Firefox, ECMA CLR
   *
   * This term has [2] labels and comments, in the languages [en, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  platform: _NS("platform"),

  /**
   * Description of target user base
   *
   * This term has [2] labels and comments, in the languages [en, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  audience: _NS("audience"),

  /**
   * URI of a blog related to a project
   *
   * This term has [2] labels and comments, in the languages [en, pt].
   *
   * Defined by the vocabulary: http://usefulinc.com/ns/doap#
   */
  blog: _NS("blog"),

}

export default DOAP;
