// @inrupt/vocab-common-rdf/GeneratedVocab/FOAF.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * [Generator provided] - Friend of a friend, v0.99
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://xmlns.com/foaf/0.1/";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * [Generator provided] - Friend of a friend, v0.99
 */
const FOAF = {
  PREFIX: "foaf",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "foaf": "http://xmlns.com/foaf/0.1/" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * An agent (eg. person, group, software or physical artifact).
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  Agent: _NS("Agent"),

  /**
   * A document.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  Document: _NS("Document"),

  /**
   * An organization.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  Organization: _NS("Organization"),

  /**
   * A project (a collective endeavour of some kind).
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  Project: _NS("Project"),

  /**
   * A person.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  Person: _NS("Person"),

  /**
   * A class of Agents.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  Group: _NS("Group"),

  /**
   * An image.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  Image: _NS("Image"),

  /**
   * A foaf:LabelProperty is any RDF property with texual values that serve as labels.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  LabelProperty: _NS("LabelProperty"),

  /**
   * An online account.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  OnlineAccount: _NS("OnlineAccount"),

  /**
   * An online chat account.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  OnlineChatAccount: _NS("OnlineChatAccount"),

  /**
   * An online e-commerce account.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  OnlineEcommerceAccount: _NS("OnlineEcommerceAccount"),

  /**
   * An online gaming account.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  OnlineGamingAccount: _NS("OnlineGamingAccount"),

  /**
   * A personal profile RDF document.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  PersonalProfileDocument: _NS("PersonalProfileDocument"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * Indicates an account held by this agent.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  account: _NS("account"),

  /**
   * Indicates the name (identifier) associated with this online account.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  accountName: _NS("accountName"),

  /**
   * Indicates a homepage of the service provide for this online account.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  accountServiceHomepage: _NS("accountServiceHomepage"),

  /**
   * The age in years of some agent.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  age: _NS("age"),

  /**
   * An AIM chat ID
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  aimChatID: _NS("aimChatID"),

  /**
   * A short informal nickname characterising an agent (includes login identifiers, IRC and other chat nicknames).
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  nick: _NS("nick"),

  /**
   * A location that something is based near, for some broadly human notion of near.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  based_near: _NS("based_near"),

  /**
   * The birthday of this Agent, represented in mm-dd string form, eg. '12-31'.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  birthday: _NS("birthday"),

  /**
   * A current project this person works on.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  currentProject: _NS("currentProject"),

  /**
   * A depiction of some thing.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  depiction: _NS("depiction"),

  /**
   * A thing depicted in this representation.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  depicts: _NS("depicts"),

  /**
   * A checksum for the DNA of some thing. Joke.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  dnaChecksum: _NS("dnaChecksum"),

  /**
   * The family name of some person.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  familyName: _NS("familyName"),

  /**
   * The family name of some person.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  family_name: _NS("family_name"),

  /**
   * The first name of a person.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  firstName: _NS("firstName"),

  /**
   * The underlying or 'focal' entity associated with some SKOS-described concept.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  focus: _NS("focus"),

  /**
   * An organization funding a project or person.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  fundedBy: _NS("fundedBy"),

  /**
   * A textual geekcode for this person, see http://www.geekcode.com/geek.html
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  geekcode: _NS("geekcode"),

  /**
   * The gender of this Agent (typically but not necessarily 'male' or 'female').
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  gender: _NS("gender"),

  /**
   * The given name of some person.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  givenName: _NS("givenName"),

  /**
   * The given name of some person.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  givenname: _NS("givenname"),

  /**
   * Indicates an account held by this agent.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  holdsAccount: _NS("holdsAccount"),

  /**
   * A homepage for some thing.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  homepage: _NS("homepage"),

  /**
   * A page or document about this thing.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  page: _NS("page"),

  /**
   * A document that this thing is the primary topic of.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  isPrimaryTopicOf: _NS("isPrimaryTopicOf"),

  /**
   * A topic of some page or document.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  topic: _NS("topic"),

  /**
   * The primary topic of some page or document.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  primaryTopic: _NS("primaryTopic"),

  /**
   * An ICQ chat ID
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  icqChatID: _NS("icqChatID"),

  /**
   * An image that can be used to represent some thing (ie. those depictions which are particularly representative of something, eg. one's photo on a homepage).
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  img: _NS("img"),

  /**
   * A page about a topic of interest to this person.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  interest: _NS("interest"),

  /**
   * A jabber ID for something.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  jabberID: _NS("jabberID"),

  /**
   * A person known by this person (indicating some level of reciprocated interaction between the parties).
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  knows: _NS("knows"),

  /**
   * The last name of a person.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  lastName: _NS("lastName"),

  /**
   * A logo representing some thing.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  logo: _NS("logo"),

  /**
   * Something that was made by this agent.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  made: _NS("made"),

  /**
   * An agent that made this thing.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  maker: _NS("maker"),

  /**
   * A personal mailbox, ie. an Internet mailbox associated with exactly one owner, the first owner of this mailbox. This is a 'static inverse functional property', in that there is (across time and change) at most one individual that ever has any particular value for foaf:mbox.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  mbox: _NS("mbox"),

  /**
   * The sha1sum of the URI of an Internet mailbox associated with exactly one owner, the first owner of the mailbox.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  mbox_sha1sum: _NS("mbox_sha1sum"),

  /**
   * Indicates a member of a Group
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  member: _NS("member"),

  /**
   * Indicates the class of individuals that are a member of a Group
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  membershipClass: _NS("membershipClass"),

  /**
   * An MSN chat ID
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  msnChatID: _NS("msnChatID"),

  /**
   * A Myers Briggs (MBTI) personality classification.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  myersBriggs: _NS("myersBriggs"),

  /**
   * A name for some thing.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  name: _NS("name"),

  /**
   * An OpenID for an Agent.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  openid: _NS("openid"),

  /**
   * A project this person has previously worked on.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  pastProject: _NS("pastProject"),

  /**
   * A phone, specified using fully qualified tel: URI scheme (refs: http://www.w3.org/Addressing/schemes.html#tel).
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  phone: _NS("phone"),

  /**
   * A .plan comment, in the tradition of finger and '.plan' files.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  plan: _NS("plan"),

  /**
   * A link to the publications of this person.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  publications: _NS("publications"),

  /**
   * A homepage of a school attended by the person.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  schoolHomepage: _NS("schoolHomepage"),

  /**
   * A sha1sum hash, in hex.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  sha1: _NS("sha1"),

  /**
   * A Skype ID
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  skypeID: _NS("skypeID"),

  /**
   * A string expressing what the user is happy for the general public (normally) to know about their current activity.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  status: _NS("status"),

  /**
   * The surname of some person.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  surname: _NS("surname"),

  /**
   * A theme.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  theme: _NS("theme"),

  /**
   * A derived thumbnail image.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  thumbnail: _NS("thumbnail"),

  /**
   * A tipjar document for this agent, describing means for payment and reward.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  tipjar: _NS("tipjar"),

  /**
   * Title (Mr, Mrs, Ms, Dr. etc)
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  title: _NS("title"),

  /**
   * A thing of interest to this person.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  topic_interest: _NS("topic_interest"),

  /**
   * A weblog of some thing (whether person, group, company etc.).
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  weblog: _NS("weblog"),

  /**
   * A work info homepage of some person; a page about their work for some organization.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  workInfoHomepage: _NS("workInfoHomepage"),

  /**
   * A workplace homepage of some person; the homepage of an organization they work for.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  workplaceHomepage: _NS("workplaceHomepage"),

  /**
   * A Yahoo chat ID
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://xmlns.com/foaf/0.1/
   */
  yahooChatID: _NS("yahooChatID"),

}

export default FOAF;
