// @inrupt/vocab-common-rdf/GeneratedVocab/HTTPH_INRUPT.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * This vocabulary is a one deemed to contain all HTTP headers.
The rdflib.js library uses it as for relationships between an HTTP response
and the  content (value) of the HTTP header in that response.     The ontology
is deemed to contain one property for every HTTP header, standard or not.
The property in this ontology has a localname which is the header field converted
to lower case.
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://www.w3.org/2007/ont/httph#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * This vocabulary is a one deemed to contain all HTTP headers.
The rdflib.js library uses it as for relationships between an HTTP response
and the  content (value) of the HTTP header in that response.     The ontology
is deemed to contain one property for every HTTP header, standard or not.
The property in this ontology has a localname which is the header field converted
to lower case.
 */
const HTTPH_INRUPT = {
  PREFIX: "httph-inrupt",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "httph-inrupt": "http://www.w3.org/2007/ont/httph#" },
  NS: _NS,

  // *******************
  // All the Properties.
  // *******************

  /**
   * See the HTTP specification
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2007/ont/httph
   */
  content_type: _NS("content-type"),

  /**
   * See the HTTP specification
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2007/ont/httph#
   */
  accept: _NS("accept"),

  /**
   * See the HTTP specification
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2007/ont/httph#
   */
  accept_language: _NS("accept-language"),

}

export default HTTPH_INRUPT;
