// @inrupt/vocab-common-rdf/GeneratedVocab/HTTP.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * A namespace for describing HTTP messages (http://www.w3.org/Protocols/rfc2616/rfc2616.html)
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://www.w3.org/2011/http#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * A namespace for describing HTTP messages (http://www.w3.org/Protocols/rfc2616/rfc2616.html)
 */
const HTTP = {
  PREFIX: "http",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "http": "http://www.w3.org/2011/http#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * A connection used for HTTP transfer.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  Connection: _NS("Connection"),

  /**
   * An entity header in an HTTP message.
   *
   * This term provides descriptions only in English.
   */
  EntityHeader: _NS("EntityHeader"),

  /**
   * A header in an HTTP message.
   *
   * This term provides descriptions only in English.
   */
  MessageHeader: _NS("MessageHeader"),

  /**
   * A general header in an HTTP message.
   *
   * This term provides descriptions only in English.
   */
  GeneralHeader: _NS("GeneralHeader"),

  /**
   * A part of a deconstructed header value.
   *
   * This term provides descriptions only in English.
   */
  HeaderElement: _NS("HeaderElement"),

  /**
   * A header name.
   *
   * This term provides descriptions only in English.
   */
  HeaderName: _NS("HeaderName"),

  /**
   * An HTTP message.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  Message: _NS("Message"),

  /**
   * The HTTP method used for the request.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  Method: _NS("Method"),

  /**
   * A parameter for a part of a header value.
   *
   * This term provides descriptions only in English.
   */
  Parameter: _NS("Parameter"),

  /**
   * An HTTP request.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  Request: _NS("Request"),

  /**
   * A header in an HTTP request message.
   *
   * This term provides descriptions only in English.
   */
  RequestHeader: _NS("RequestHeader"),

  /**
   * An HTTP response.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  Response: _NS("Response"),

  /**
   * A header in an HTTP response message.
   *
   * This term provides descriptions only in English.
   */
  ResponseHeader: _NS("ResponseHeader"),

  /**
   * The status code of an HTTP response.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  StatusCode: _NS("StatusCode"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * The absolute request URI.
   *
   * This term provides descriptions only in English.
   */
  absoluteURI: _NS("absoluteURI"),

  /**
   * The request URI of an HTTP request.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  requestURI: _NS("requestURI"),

  /**
   * The absolute path sort of request URI.
   *
   * This term provides descriptions only in English.
   */
  absolutePath: _NS("absolutePath"),

  /**
   * The authority sort of request URI.
   *
   * This term provides descriptions only in English.
   */
  authority: _NS("authority"),

  /**
   * The entity body of an HTTP message.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  body: _NS("body"),

  /**
   * The authority of a connection used for the HTTP transfer.
   *
   * This term provides descriptions only in English.
   */
  connectionAuthority: _NS("connectionAuthority"),

  /**
   * The name of a header element.
   *
   * This term provides descriptions only in English.
   */
  elementName: _NS("elementName"),

  /**
   * The value of a header element.
   *
   * This term provides descriptions only in English.
   */
  elementValue: _NS("elementValue"),

  /**
   * The name of an HTTP header field.
   *
   * This term provides descriptions only in English.
   */
  fieldName: _NS("fieldName"),

  /**
   * The value of an HTTP header field.
   *
   * This term provides descriptions only in English.
   */
  fieldValue: _NS("fieldValue"),

  /**
   * The deconstructed parts of an HTTP header value.
   *
   * This term provides descriptions only in English.
   */
  headerElements: _NS("headerElements"),

  /**
   * The name of an HTTP header.
   *
   * This term provides descriptions only in English.
   */
  hdrName: _NS("hdrName"),

  /**
   * The headers in an HTTP message.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  headers: _NS("headers"),

  /**
   * The HTTP version of an HTTP message.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  httpVersion: _NS("httpVersion"),

  /**
   * The HTTP method used for the HTTP request.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  mthd: _NS("mthd"),

  /**
   * The HTTP method name used for the HTTP request.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  methodName: _NS("methodName"),

  /**
   * The name of a parameter in a part of a deconstructed HTTP header value.
   *
   * This term provides descriptions only in English.
   */
  paramName: _NS("paramName"),

  /**
   * The parameters in a part of a deconstructed HTTP header value.
   *
   * This term provides descriptions only in English.
   */
  params: _NS("params"),

  /**
   * The value of a parameter in a part of a deconstructed HTTP header value.
   *
   * This term provides descriptions only in English.
   */
  paramValue: _NS("paramValue"),

  /**
   * The reason phrase (status text) of an HTTP response.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  reasonPhrase: _NS("reasonPhrase"),

  /**
   * The HTTP requests made via a connection.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  requests: _NS("requests"),

  /**
   * The HTTP response sent in answer to an HTTP request.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  resp: _NS("resp"),

  /**
   * The status code of an HTTP response.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  sc: _NS("sc"),

  /**
   * The status code number.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  statusCodeNumber: _NS("statusCodeNumber"),

  /**
   * The status code value of an HTTP response.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.ietf.org/rfc/rfc2616.txt
   */
  statusCodeValue: _NS("statusCodeValue"),

}

export default HTTP;
