// @inrupt/vocab-common-rdf/GeneratedVocab/ICAL.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * [Generator provided] - iCal - a vocabulary for description of events and calendars
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "https://www.w3.org/2002/12/cal/ical#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * [Generator provided] - iCal - a vocabulary for description of events and calendars
 */
const ICAL = {
  PREFIX: "ical",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "ical": "https://www.w3.org/2002/12/cal/ical#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * Provide a grouping of component properties that define an alarm.
   *
   * This term provides descriptions only with no explicit locale.
   */
  Valarm: _NS("Valarm"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  Value_DURATION: _NS("Value_DURATION"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  Value_CAL_ADDRESS: _NS("Value_CAL-ADDRESS"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  Value_DATE: _NS("Value_DATE"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  Value_DATE_TIME: _NS("Value_DATE-TIME"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  Value_PERIOD: _NS("Value_PERIOD"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  Value_RECUR: _NS("Value_RECUR"),

  /**
   * Provide a grouping of component properties that describe an event.
   *
   * This term provides descriptions only with no explicit locale.
   */
  Vevent: _NS("Vevent"),

  /**
   * Provide a grouping of component properties that describe either a request for free/busy time, describe a response to a request for free/busy time or describe a published set of busy time.
   *
   * This term provides descriptions only with no explicit locale.
   */
  Vfreebusy: _NS("Vfreebusy"),

  /**
   * Provide a grouping of component properties that describe a journal entry.
   *
   * This term provides descriptions only with no explicit locale.
   */
  Vjournal: _NS("Vjournal"),

  /**
   * Provide a grouping of component properties that defines a time zone.
   *
   * This term provides descriptions only with no explicit locale.
   */
  Vtimezone: _NS("Vtimezone"),

  /**
   * Provide a grouping of calendar properties that describe a to-do.
   *
   * This term provides descriptions only with no explicit locale.
   */
  Vtodo: _NS("Vtodo"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * 
	    default value type: URI
   *
   * This term provides descriptions only with no explicit locale.
   */
  attach: _NS("attach"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  description: _NS("description"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  summary: _NS("summary"),

  /**
   * 
	    value type: DURATION
   *
   * This term provides descriptions only with no explicit locale.
   */
  duration: _NS("duration"),

  /**
   * 
	    value type: CAL-ADDRESS
   *
   * This term provides descriptions only with no explicit locale.
   */
  attendee: _NS("attendee"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  action: _NS("action"),

  /**
   * 
	    value type: INTEGER
   *
   * This term provides descriptions only with no explicit locale.
   */
  repeat: _NS("repeat"),

  /**
   * 
	    default value type: DURATION
   *
   * This term provides descriptions only with no explicit locale.
   */
  trigger: _NS("trigger"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  categories: _NS("categories"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  class: _NS("class"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  comment: _NS("comment"),

  /**
   * This property specifies information related to the global position for the activity specified by a calendar component.
   *
   * This term provides descriptions only with no explicit locale.
   */
  geo: _NS("geo"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  location: _NS("location"),

  /**
   * 
	    value type: INTEGER
   *
   * This term provides descriptions only with no explicit locale.
   */
  priority: _NS("priority"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  resources: _NS("resources"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  status: _NS("status"),

  /**
   * 
	    default value type: DATE-TIME
   *
   * This term provides descriptions only with no explicit locale.
   */
  dtend: _NS("dtend"),

  /**
   * 
	    default value type: DATE-TIME
   *
   * This term provides descriptions only with no explicit locale.
   */
  dtstart: _NS("dtstart"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  transp: _NS("transp"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  contact: _NS("contact"),

  /**
   * 
	    value type: CAL-ADDRESS
   *
   * This term provides descriptions only with no explicit locale.
   */
  organizer: _NS("organizer"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  relatedTo: _NS("relatedTo"),

  /**
   * 
	    value type: URI
   *
   * This term provides descriptions only with no explicit locale.
   */
  url: _NS("url"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  uid: _NS("uid"),

  /**
   * 
	    value type: RECUR
   *
   * This term provides descriptions only with no explicit locale.
   */
  exrule: _NS("exrule"),

  /**
   * 
	    default value type: DATE-TIME
   *
   * This term provides descriptions only with no explicit locale.
   */
  rdate: _NS("rdate"),

  /**
   * 
	    value type: RECUR
   *
   * This term provides descriptions only with no explicit locale.
   */
  rrule: _NS("rrule"),

  /**
   * 
	    value type: DATE-TIME
   *
   * This term provides descriptions only with no explicit locale.
   */
  created: _NS("created"),

  /**
   * 
	    value type: DATE-TIME
   *
   * This term provides descriptions only with no explicit locale.
   */
  dtstamp: _NS("dtstamp"),

  /**
   * 
	    value type: DATE-TIME
   *
   * This term provides descriptions only with no explicit locale.
   */
  lastModified: _NS("lastModified"),

  /**
   * 
	    value type: integer
   *
   * This term provides descriptions only with no explicit locale.
   */
  sequence: _NS("sequence"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  requestStatus: _NS("requestStatus"),

  /**
   * 
	    value type: PERIOD
   *
   * This term provides descriptions only with no explicit locale.
   */
  freebusy: _NS("freebusy"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  tzid: _NS("tzid"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  tzname: _NS("tzname"),

  /**
   * 
	    value type: UTC-OFFSET
   *
   * This term provides descriptions only with no explicit locale.
   */
  tzoffsetfrom: _NS("tzoffsetfrom"),

  /**
   * 
	    value type: UTC-OFFSET
   *
   * This term provides descriptions only with no explicit locale.
   */
  tzoffsetto: _NS("tzoffsetto"),

  /**
   * 
	    value type: URI
   *
   * This term provides descriptions only with no explicit locale.
   */
  tzurl: _NS("tzurl"),

  /**
   * 
	    value type: INTEGER
   *
   * This term provides descriptions only with no explicit locale.
   */
  percentComplete: _NS("percentComplete"),

  /**
   * 
	    value type: DATE-TIME
   *
   * This term provides descriptions only with no explicit locale.
   */
  completed: _NS("completed"),

  /**
   * 
	    default value type: DATE-TIME
   *
   * This term provides descriptions only with no explicit locale.
   */
  due: _NS("due"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  X_: _NS("X-"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  calscale: _NS("calscale"),

  /**
   * 
	    default value type: DATE-TIME
   *
   * This term provides descriptions only with no explicit locale.
   */
  exdate: _NS("exdate"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  method: _NS("method"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  prodid: _NS("prodid"),

  /**
   * 
	    default value type: DATE-TIME
   *
   * This term provides descriptions only with no explicit locale.
   */
  recurrenceId: _NS("recurrenceId"),

  /**
   * 
	    value type: TEXT
   *
   * This term provides descriptions only with no explicit locale.
   */
  version: _NS("version"),

}

export default ICAL;
