// @inrupt/vocab-common-rdf/GeneratedVocab/OLO.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * [Generator provided] - OLO (Ordered List Ontology) - for describing ordered lists
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://purl.org/ontology/olo/core#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * [Generator provided] - OLO (Ordered List Ontology) - for describing ordered lists
 */
const OLO = {
  PREFIX: "olo",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "olo": "http://purl.org/ontology/olo/core#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * An ordered list with a given length an indexed items.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/ontology/olo/core#
   */
  OrderedList: _NS("OrderedList"),

  /**
   * A slot in an ordered list with a fixed index.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/ontology/olo/core#
   */
  Slot: _NS("Slot"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * The length of an ordered list.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/ontology/olo/core#
   */
  length: _NS("length"),

  /**
   * Associates the next slot in an ordered list.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/ontology/olo/core#
   */
  next: _NS("next"),

  /**
   * An index of a slot in an ordered list.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/ontology/olo/core#
   */
  index: _NS("index"),

  /**
   * An ordered list of an slot.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/ontology/olo/core#
   */
  ordered_list: _NS("ordered_list"),

  /**
   * Associates the previous slot in an ordered list
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/ontology/olo/core#
   */
  previous: _NS("previous"),

  /**
   * A slot in an ordered list.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/ontology/olo/core#
   */
  slot: _NS("slot"),

  /**
   * An item of a slot in an ordered list.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://purl.org/ontology/olo/core#
   */
  item: _NS("item"),

}

export default OLO;
