// @inrupt/vocab-common-rdf/GeneratedVocab/OWL.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * 
  This ontology partially describes the built-in classes and
  properties that together form the basis of the RDF/XML syntax of OWL 2.
  The content of this ontology is based on Tables 6.1 and 6.2
  in Section 6.4 of the OWL 2 RDF-Based Semantics specification,
  available at http://www.w3.org/TR/owl2-rdf-based-semantics/.
  Please note that those tables do not include the different annotations
  (labels, comments and rdfs:isDefinedBy links) used in this file.
  Also note that the descriptions provided in this ontology do not
  provide a complete and correct formal description of either the syntax
  or the semantics of the introduced terms (please see the OWL 2
  recommendations for the complete and normative specifications).
  Furthermore, the information provided by this ontology may be
  misleading if not used with care. This ontology SHOULD NOT be imported
  into OWL ontologies. Importing this file into an OWL 2 DL ontology
  will cause it to become an OWL 2 Full ontology and may have other,
  unexpected, consequences.
   
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://www.w3.org/2002/07/owl#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * 
  This ontology partially describes the built-in classes and
  properties that together form the basis of the RDF/XML syntax of OWL 2.
  The content of this ontology is based on Tables 6.1 and 6.2
  in Section 6.4 of the OWL 2 RDF-Based Semantics specification,
  available at http://www.w3.org/TR/owl2-rdf-based-semantics/.
  Please note that those tables do not include the different annotations
  (labels, comments and rdfs:isDefinedBy links) used in this file.
  Also note that the descriptions provided in this ontology do not
  provide a complete and correct formal description of either the syntax
  or the semantics of the introduced terms (please see the OWL 2
  recommendations for the complete and normative specifications).
  Furthermore, the information provided by this ontology may be
  misleading if not used with care. This ontology SHOULD NOT be imported
  into OWL ontologies. Importing this file into an OWL 2 DL ontology
  will cause it to become an OWL 2 Full ontology and may have other,
  unexpected, consequences.
   
 */
const OWL = {
  PREFIX: "owl",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "owl": "http://www.w3.org/2002/07/owl#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * The class of ontologies.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  Ontology: _NS("Ontology"),

  /**
   * The class of ontology properties.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  OntologyProperty: _NS("OntologyProperty"),

  /**
   * The class of annotation properties.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  AnnotationProperty: _NS("AnnotationProperty"),

  /**
   * The class of collections of pairwise different individuals.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  AllDifferent: _NS("AllDifferent"),

  /**
   * The class of collections of pairwise disjoint classes.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  AllDisjointClasses: _NS("AllDisjointClasses"),

  /**
   * The class of collections of pairwise disjoint properties.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  AllDisjointProperties: _NS("AllDisjointProperties"),

  /**
   * The class of annotated annotations for which the RDF serialization consists of an annotated subject, predicate and object.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  Annotation: _NS("Annotation"),

  /**
   * The class of asymmetric properties.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  AsymmetricProperty: _NS("AsymmetricProperty"),

  /**
   * The class of object properties.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  ObjectProperty: _NS("ObjectProperty"),

  /**
   * The class of annotated axioms for which the RDF serialization consists of an annotated subject, predicate and object.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  Axiom: _NS("Axiom"),

  /**
   * The class of OWL classes.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  Class: _NS("Class"),

  /**
   * The class of OWL data ranges, which are special kinds of datatypes. Note: The use of the IRI owl:DataRange has been deprecated as of OWL 2. The IRI rdfs:Datatype SHOULD be used instead.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  DataRange: _NS("DataRange"),

  /**
   * The class of data properties.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  DatatypeProperty: _NS("DatatypeProperty"),

  /**
   * The class of deprecated classes.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  DeprecatedClass: _NS("DeprecatedClass"),

  /**
   * The class of deprecated properties.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  DeprecatedProperty: _NS("DeprecatedProperty"),

  /**
   * The class of functional properties.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  FunctionalProperty: _NS("FunctionalProperty"),

  /**
   * The class of inverse-functional properties.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  InverseFunctionalProperty: _NS("InverseFunctionalProperty"),

  /**
   * The class of irreflexive properties.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  IrreflexiveProperty: _NS("IrreflexiveProperty"),

  /**
   * The class of named individuals.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  NamedIndividual: _NS("NamedIndividual"),

  /**
   * The class of OWL individuals.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  Thing: _NS("Thing"),

  /**
   * The class of negative property assertions.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  NegativePropertyAssertion: _NS("NegativePropertyAssertion"),

  /**
   * This is the empty class.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  Nothing: _NS("Nothing"),

  /**
   * The class of reflexive properties.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  ReflexiveProperty: _NS("ReflexiveProperty"),

  /**
   * The class of property restrictions.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  Restriction: _NS("Restriction"),

  /**
   * The class of symmetric properties.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  SymmetricProperty: _NS("SymmetricProperty"),

  /**
   * The class of transitive properties.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  TransitiveProperty: _NS("TransitiveProperty"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * The annotation property that provides version information for an ontology or another OWL construct.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  versionInfo: _NS("versionInfo"),

  /**
   * The property that determines the class that a universal property restriction refers to.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  allValuesFrom: _NS("allValuesFrom"),

  /**
   * The property that determines the predicate of an annotated axiom or annotated annotation.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  annotatedProperty: _NS("annotatedProperty"),

  /**
   * The property that determines the subject of an annotated axiom or annotated annotation.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  annotatedSource: _NS("annotatedSource"),

  /**
   * The property that determines the object of an annotated axiom or annotated annotation.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  annotatedTarget: _NS("annotatedTarget"),

  /**
   * The property that determines the predicate of a negative property assertion.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  assertionProperty: _NS("assertionProperty"),

  /**
   * The annotation property that indicates that a given ontology is backward compatible with another ontology.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  backwardCompatibleWith: _NS("backwardCompatibleWith"),

  /**
   * The data property that does not relate any individual to any data value.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  bottomDataProperty: _NS("bottomDataProperty"),

  /**
   * The object property that does not relate any two individuals.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  bottomObjectProperty: _NS("bottomObjectProperty"),

  /**
   * The property that determines the cardinality of an exact cardinality restriction.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  cardinality: _NS("cardinality"),

  /**
   * The property that determines that a given class is the complement of another class.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  complementOf: _NS("complementOf"),

  /**
   * The property that determines that a given data range is the complement of another data range with respect to the data domain.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  datatypeComplementOf: _NS("datatypeComplementOf"),

  /**
   * The annotation property that indicates that a given entity has been deprecated.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  deprecated: _NS("deprecated"),

  /**
   * The property that determines that two given individuals are different.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  differentFrom: _NS("differentFrom"),

  /**
   * The property that determines that a given class is equivalent to the disjoint union of a collection of other classes.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  disjointUnionOf: _NS("disjointUnionOf"),

  /**
   * The property that determines that two given classes are disjoint.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  disjointWith: _NS("disjointWith"),

  /**
   * The property that determines the collection of pairwise different individuals in a owl:AllDifferent axiom.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  distinctMembers: _NS("distinctMembers"),

  /**
   * The property that determines that two given classes are equivalent, and that is used to specify datatype definitions.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  equivalentClass: _NS("equivalentClass"),

  /**
   * The property that determines that two given properties are equivalent.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  equivalentProperty: _NS("equivalentProperty"),

  /**
   * The property that determines the collection of properties that jointly build a key.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  hasKey: _NS("hasKey"),

  /**
   * The property that determines the property that a self restriction refers to.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  hasSelf: _NS("hasSelf"),

  /**
   * The property that determines the individual that a has-value restriction refers to.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  hasValue: _NS("hasValue"),

  /**
   * The annotation property that indicates that a given ontology is incompatible with another ontology.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  incompatibleWith: _NS("incompatibleWith"),

  /**
   * The property that determines the collection of classes or data ranges that build an intersection.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  intersectionOf: _NS("intersectionOf"),

  /**
   * The property that determines that two given properties are inverse.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  inverseOf: _NS("inverseOf"),

  /**
   * The property that determines the cardinality of a maximum cardinality restriction.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  maxCardinality: _NS("maxCardinality"),

  /**
   * The property that determines the cardinality of a maximum qualified cardinality restriction.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  maxQualifiedCardinality: _NS("maxQualifiedCardinality"),

  /**
   * The property that determines the collection of members in either a owl:AllDifferent, owl:AllDisjointClasses or owl:AllDisjointProperties axiom.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  members: _NS("members"),

  /**
   * The property that determines the cardinality of a minimum cardinality restriction.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  minCardinality: _NS("minCardinality"),

  /**
   * The property that determines the cardinality of a minimum qualified cardinality restriction.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  minQualifiedCardinality: _NS("minQualifiedCardinality"),

  /**
   * The property that determines the class that a qualified object cardinality restriction refers to.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  onClass: _NS("onClass"),

  /**
   * The property that determines the data range that a qualified data cardinality restriction refers to.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  onDataRange: _NS("onDataRange"),

  /**
   * The property that determines the datatype that a datatype restriction refers to.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  onDatatype: _NS("onDatatype"),

  /**
   * The property that determines the collection of individuals or data values that build an enumeration.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  oneOf: _NS("oneOf"),

  /**
   * The property that determines the n-tuple of properties that a property restriction on an n-ary data range refers to.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  onProperties: _NS("onProperties"),

  /**
   * The property that determines the property that a property restriction refers to.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  onProperty: _NS("onProperty"),

  /**
   * The annotation property that indicates the predecessor ontology of a given ontology.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  priorVersion: _NS("priorVersion"),

  /**
   * The property that determines the n-tuple of properties that build a sub property chain of a given property.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  propertyChainAxiom: _NS("propertyChainAxiom"),

  /**
   * The property that determines that two given properties are disjoint.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  propertyDisjointWith: _NS("propertyDisjointWith"),

  /**
   * The property that determines the cardinality of an exact qualified cardinality restriction.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  qualifiedCardinality: _NS("qualifiedCardinality"),

  /**
   * The property that determines that two given individuals are equal.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  sameAs: _NS("sameAs"),

  /**
   * The property that determines the class that an existential property restriction refers to.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  someValuesFrom: _NS("someValuesFrom"),

  /**
   * The property that determines the subject of a negative property assertion.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  sourceIndividual: _NS("sourceIndividual"),

  /**
   * The property that determines the object of a negative object property assertion.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  targetIndividual: _NS("targetIndividual"),

  /**
   * The property that determines the value of a negative data property assertion.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  targetValue: _NS("targetValue"),

  /**
   * The data property that relates every individual to every data value.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  topDataProperty: _NS("topDataProperty"),

  /**
   * The object property that relates every two individuals.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  topObjectProperty: _NS("topObjectProperty"),

  /**
   * The property that determines the collection of classes or data ranges that build a union.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  unionOf: _NS("unionOf"),

  /**
   * The property that determines the collection of facet-value pairs that define a datatype restriction.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  withRestrictions: _NS("withRestrictions"),

}

export default OWL;
