// @inrupt/vocab-common-rdf/GeneratedVocab/POSIX.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * A vocabulary for the basic POSIX terms.
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://www.w3.org/ns/posix/stat#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * A vocabulary for the basic POSIX terms.
 */
const POSIX = {
  PREFIX: "posix",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "posix": "http://www.w3.org/ns/posix/stat#" },
  NS: _NS,

  // *******************
  // All the Properties.
  // *******************

  /**
   * Time of last access
   *
   * This term provides descriptions only in English.
   */
  atime: _NS("atime"),

  /**
   * Blocksize for file system I/O
   *
   * This term provides descriptions only in English.
   */
  blksize: _NS("blksize"),

  /**
   * Number of 512B blocks allocated
   *
   * This term provides descriptions only in English.
   */
  blocks: _NS("blocks"),

  /**
   * Time of last status change
   *
   * This term provides descriptions only in English.
   */
  ctime: _NS("ctime"),

  /**
   * ID of device containing file
   *
   * This term provides descriptions only in English.
   */
  dev: _NS("dev"),

  /**
   * Group ID of owner
   *
   * This term provides descriptions only in English.
   */
  gid: _NS("gid"),

  /**
   * inode number
   *
   * This term provides descriptions only in English.
   */
  ino: _NS("ino"),

  /**
   * Protection
   *
   * This term provides descriptions only in English.
   */
  mode: _NS("mode"),

  /**
   * Time of last modification
   *
   * This term provides descriptions only in English.
   */
  mtime: _NS("mtime"),

  /**
   * Number of hard links
   *
   * This term provides descriptions only in English.
   */
  nlink: _NS("nlink"),

  /**
   * Device ID (if special file)
   *
   * This term provides descriptions only in English.
   */
  rdev: _NS("rdev"),

  /**
   * Total size, in bytes
   *
   * This term provides descriptions only in English.
   */
  size: _NS("size"),

  /**
   * User ID of owner
   *
   * This term provides descriptions only in English.
   */
  uid: _NS("uid"),

}

export default POSIX;
