// @inrupt/vocab-common-rdf/GeneratedVocab/RDFS.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * [Generator provided] - RDFS (RDF Schema) - a data-modelling vocabulary for RDF data
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://www.w3.org/2000/01/rdf-schema#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * [Generator provided] - RDFS (RDF Schema) - a data-modelling vocabulary for RDF data
 */
const RDFS = {
  PREFIX: "rdfs",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "rdfs": "http://www.w3.org/2000/01/rdf-schema#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * The class resource, everything.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2000/01/rdf-schema#
   */
  Resource: _NS("Resource"),

  /**
   * The class of classes.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2000/01/rdf-schema#
   */
  Class: _NS("Class"),

  /**
   * The class of literal values, eg. textual strings and integers.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2000/01/rdf-schema#
   */
  Literal: _NS("Literal"),

  /**
   * The class of RDF containers.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2000/01/rdf-schema#
   */
  Container: _NS("Container"),

  /**
   * The class of container membership properties, rdf:_1, rdf:_2, ...,
                    all of which are sub-properties of 'member'.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2000/01/rdf-schema#
   */
  ContainerMembershipProperty: _NS("ContainerMembershipProperty"),

  /**
   * The class of RDF datatypes.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2000/01/rdf-schema#
   */
  Datatype: _NS("Datatype"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * Further information about the subject resource.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2000/01/rdf-schema#
   */
  seeAlso: _NS("seeAlso"),

  /**
   * The defininition of the subject resource.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2000/01/rdf-schema#
   */
  isDefinedBy: _NS("isDefinedBy"),

  /**
   * A human-readable name for the subject.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2000/01/rdf-schema#
   */
  label: _NS("label"),

  /**
   * A description of the subject resource.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2000/01/rdf-schema#
   */
  comment: _NS("comment"),

  /**
   * A range of the subject property.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2000/01/rdf-schema#
   */
  range: _NS("range"),

  /**
   * A domain of the subject property.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2000/01/rdf-schema#
   */
  domain: _NS("domain"),

  /**
   * The subject is a subclass of a class.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2000/01/rdf-schema#
   */
  subClassOf: _NS("subClassOf"),

  /**
   * The subject is a subproperty of a property.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2000/01/rdf-schema#
   */
  subPropertyOf: _NS("subPropertyOf"),

  /**
   * A member of the subject resource.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2000/01/rdf-schema#
   */
  member: _NS("member"),

}

export default RDFS;
