// @inrupt/vocab-common-rdf/GeneratedVocab/RLOG.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * Inrupt-created copy of RLOG, needed due to Server 500 errors resolving original uni-leipzig.org!
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * Inrupt-created copy of RLOG, needed due to Server 500 errors resolving original uni-leipzig.org!
 */
const RLOG = {
  PREFIX: "rlog",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "rlog": "http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * look here: http://logging.apache.org/log4j/1.2/apidocs/org/apache/log4j/Level.html 
   *
   * This term provides descriptions only in English.
   */
  Level: _NS("Level"),

  /**
   * An entry in a log.
   *
   * This term provides descriptions only in English.
   */
  Entry: _NS("Entry"),

  /**
   * A status code which can occur in an application. Instances of this class must put the name of the status as rdfs:label and describe it with rdfs:comment.
   *
   * This term provides descriptions only in English.
   */
  StatusCode: _NS("StatusCode"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * The ALL has the lowest possible rank and is intended to turn on all logging.
   *
   * This term provides descriptions only in English.
   */
  ALL: _NS("ALL"),

  /**
   * Attention: the ids here can change, they are just used to define an order over the levels!
   *
   * This term provides descriptions only in English.
   */
  priority: _NS("priority"),

  /**
   * The DEBUG Level designates fine-grained informational events that are most useful to debug an application.
   *
   * This term provides descriptions only in English.
   */
  DEBUG: _NS("DEBUG"),

  /**
   * The ERROR level designates error events that might still allow the application to continue running.
   *
   * This term provides descriptions only in English.
   */
  ERROR: _NS("ERROR"),

  /**
   * The FATAL level designates very severe error events that will presumably lead the application to abort.
   *
   * This term provides descriptions only in English.
   */
  FATAL: _NS("FATAL"),

  /**
   * The INFO level designates informational messages that highlight the progress of the application at coarse-grained level.
   *
   * This term provides descriptions only in English.
   */
  INFO: _NS("INFO"),

  /**
   * The OFF has the highest possible rank and is intended to turn off logging.
   *
   * This term provides descriptions only in English.
   */
  OFF: _NS("OFF"),

  /**
   * The TRACE Level designates finer-grained informational events than the DEBUG.
   *
   * This term provides descriptions only in English.
   */
  TRACE: _NS("TRACE"),

  /**
   * The WARN level designates potentially harmful situations.
   *
   * This term provides descriptions only in English.
   */
  WARN: _NS("WARN"),

  /**
   * Outputs the fully qualified class name of the caller issuing the logging request.
   *
   * This term provides descriptions only in English.
   */
  className: _NS("className"),

  /**
   * The numerical value of the code, e.g. 200, 404 or 42.
   *
   * This term provides descriptions only in English.
   */
  codeId: _NS("codeId"),

  /**
   * Logging datetime (or just date)
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  date: _NS("date"),

  /**
   * Links a log message to a status code.
   *
   * This term provides descriptions only in English.
   */
  hasCode: _NS("hasCode"),

  /**
   * logLevel
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  level: _NS("level"),

  /**
   * Logging message
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  message: _NS("message"),

  /**
   * If we mix normal and log output, this can be used to refer to the resource in the RDF the error is connected to. 
   *
   * This term provides descriptions only in English.
   */
  resource: _NS("resource"),

}

export default RLOG;
