// @inrupt/vocab-common-rdf/GeneratedVocab/SDMX_DIMENSION.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * [Generator provided] - SDMX Dimension Vocabulary
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://purl.org/linked-data/sdmx/2009/dimension#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * [Generator provided] - SDMX Dimension Vocabulary
 */
const SDMX_DIMENSION = {
  PREFIX: "sdmx-dimension",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "sdmx-dimension": "http://purl.org/linked-data/sdmx/2009/dimension#" },
  NS: _NS,

  // *******************
  // All the Properties.
  // *******************

  /**
   * The length of time that a person has lived or a thing has existed.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://sdmx.org/wp-content/uploads/01_sdmx_cog_annex_1_cdc_2009.pdf
   */
  age: _NS("age"),

  /**
   * Legal, conjugal status of each individual in relation to the marriage laws or customs of the country.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://sdmx.org/wp-content/uploads/01_sdmx_cog_annex_1_cdc_2009.pdf
   */
  civilStatus: _NS("civilStatus"),

  /**
   * Monetary denomination of the object being measured.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://sdmx.org/wp-content/uploads/01_sdmx_cog_annex_1_cdc_2009.pdf
   */
  currency: _NS("currency"),

  /**
   * The highest level of an educational programme the person has successfully completed.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://sdmx.org/wp-content/uploads/01_sdmx_cog_annex_1_cdc_2009.pdf
   */
  educationLev: _NS("educationLev"),

  /**
   * The time interval at which observations occur over a given time period.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://sdmx.org/wp-content/uploads/01_sdmx_cog_annex_1_cdc_2009.pdf
   */
  freq: _NS("freq"),

  /**
   * Job or position held by an individual who performs a set of tasks and duties.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://sdmx.org/wp-content/uploads/01_sdmx_cog_annex_1_cdc_2009.pdf
   */
  occupation: _NS("occupation"),

  /**
   * The country or geographic area to which the measured statistical phenomenon relates.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://sdmx.org/wp-content/uploads/01_sdmx_cog_annex_1_cdc_2009.pdf
   */
  refArea: _NS("refArea"),

  /**
   * The period of time or point in time to which the measured observation is intended to refer.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://sdmx.org/wp-content/uploads/01_sdmx_cog_annex_1_cdc_2009.pdf
   */
  refPeriod: _NS("refPeriod"),

  /**
   * The state of being male or female.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://sdmx.org/wp-content/uploads/01_sdmx_cog_annex_1_cdc_2009.pdf
   */
  sex: _NS("sex"),

  /**
   * The period of time or point in time to which the measured observation refers.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: https://sdmx.org/wp-content/uploads/01_sdmx_cog_annex_1_cdc_2009.pdf
   */
  timePeriod: _NS("timePeriod"),

}

export default SDMX_DIMENSION;
