// @inrupt/vocab-common-rdf/GeneratedVocab/SKOS_XL.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * An RDF vocabulary extending SKOS and allowing the description and linking of lexical entities.
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://www.w3.org/2008/05/skos-xl#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * An RDF vocabulary extending SKOS and allowing the description and linking of lexical entities.
 */
const SKOS_XL = {
  PREFIX: "skos-xl",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "skos-xl": "http://www.w3.org/2008/05/skos-xl#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * A special class of lexical entities.
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2008/05/skos-xl
   */
  Label: _NS("Label"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * If two instances of the class skosxl:Label have the same literal form, they are not necessarily the same resource.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2008/05/skos-xl
   */
  literalForm: _NS("literalForm"),

  /**
   * If C skosxl:prefLabel L and L skosxl:literalForm V, then X skos:prefLabel V.
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://www.w3.org/2004/02/skos/core#prefLabel
   *
   * Defined by the vocabulary: http://www.w3.org/2008/05/skos-xl
   */
  prefLabel: _NS("prefLabel"),

  /**
   * If C skosxl:altLabel L and L skosxl:literalForm V, then X skos:altLabel V.
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://www.w3.org/2004/02/skos/core#altLabel
   *
   * Defined by the vocabulary: http://www.w3.org/2008/05/skos-xl
   */
  altLabel: _NS("altLabel"),

  /**
   * If C skosxl:hiddenLabel L and L skosxl:literalForm V, then C skos:hiddenLabel V.
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://www.w3.org/2004/02/skos/core#hiddenLabel
   *
   * Defined by the vocabulary: http://www.w3.org/2008/05/skos-xl
   */
  hiddenLabel: _NS("hiddenLabel"),

  /**
   * The property skosxl:labelRelation is used for representing binary ('direct') relations between instances of the class skosxl:Label.
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2008/05/skos-xl
   */
  labelRelation: _NS("labelRelation"),

}

export default SKOS_XL;
