// @inrupt/vocab-common-rdf/GeneratedVocab/UI.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * [Generator provided] - UI Ontology
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://www.w3.org/ns/ui#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * [Generator provided] - UI Ontology
 */
const UI = {
  PREFIX: "ui",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "ui": "http://www.w3.org/ns/ui#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  BooleanField: _NS("BooleanField"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  NumericField: _NS("NumericField"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  ValueField: _NS("ValueField"),

  /**
   * A form can be any type of single field, or typically a Group of several fields,
    including interspersed headings and comments.  
   *
   * The term has a description only in English, but has a mismatch between its labels and comments, with [0] labels, but [1] comment in the language [NoLocale].
   */
  Form: _NS("Form"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  Single: _NS("Single"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  Group: _NS("Group"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  Choice: _NS("Choice"),

  /**
   * A classifier allows the user to select the type of an object.
    The possible types must be subclasses of some overall class, the "category".
    (Ideally, the superclass is also set up as the disjoint union of the subclasses,
    if they are disjoint.)

    The form normally stores the resulting classes using an r:type triple,
    but a different predicate can be used if required, so the classifier field
    needs is 'property' defined too.

    If the subclass selected itself is has subclasses defined, the user can
    recursively select from them in turn, as many levels as needed.
   *
   * This term provides descriptions only with no explicit locale.
   */
  Classifier: _NS("Classifier"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  ColorField: _NS("ColorField"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  TextField: _NS("TextField"),

  /**
   * Comment
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  Comment: _NS("Comment"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  DateField: _NS("DateField"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  DateTimeField: _NS("DateTimeField"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  DecimalField: _NS("DecimalField"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  EmailField: _NS("EmailField"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  Heading: _NS("Heading"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  SingleLineTextField: _NS("SingleLineTextField"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  Options: _NS("Options"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  IntegerField: _NS("IntegerField"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  FloatField: _NS("FloatField"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  PhoneField: _NS("PhoneField"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  Multiple: _NS("Multiple"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  MultiLineTextField: _NS("MultiLineTextField"),

  /**
   * A NamedNodeURIField is like a SingleLineTextField, except
		that the value it generates is not a literal string but an RDF node with the given URI.
		Normally users should not see URIs. When they do, this a way to do it.
		
   *
   * This term provides descriptions only with no explicit locale.
   */
  NamedNodeURIField: _NS("NamedNodeURIField"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  TriStateField: _NS("TriStateField"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * parts
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  parts: _NS("parts"),

  /**
   * A form which may be used to collect information about a
    hitherto locally undocumented instance instance of this class.
   *
   * This term provides descriptions only with no explicit locale.
   */
  creationForm: _NS("creationForm"),

  /**
   * from
   *
   * This term has a label (in language [NoLocale]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  from: _NS("from"),

  /**
   * Label
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  label: _NS("label"),

  /**
   * Many fields prompt for information about a given property of the subject.
    When field is filled in, this gives which property is written into the data.
   *
   * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
   */
  property: _NS("property"),

  /**
   * The value for which this case is selected.
   *
   * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
   */
  for: _NS("for"),

  /**
   * A string for the UI to use if the user needs a longer
        prompts than just a field name, the s:label. 
   *
   * This term provides descriptions only with no explicit locale.
   */
  prompt: _NS("prompt"),

  /**
   *
   *
   * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
   */
  use: _NS("use"),

  /**
   * Must be a valid CSS color string such as one could put in
			an HTML style attribute.  This must be in the #xxxxxx form,
			(with 6 digits of lowercase hex) so that it
			can work eg with Graphviz.
			As this is just an encoded array of RGB values,
			you can do math with these, such as blending, complement, etc.
   *
   * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
   */
  Color: _NS("Color"),

  /**
   * Contents
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  contents: _NS("contents"),

  /**
   * size of field
   *
   * This term has a label (in language [NoLocale]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  size: _NS("size"),

  /**
   * Must be a valid CSS style string such as one could put in
        an HTML style attribute.  Depending on the user interface system, this can
        by given to individuals, classes or properties. It is up to a user interface
        which wants to draw on them to pick how it uses styles from which parts
        of the data it has.  For example, the style of a class may be picked
        to distinguish information about things in that class.
   *
   * This term provides descriptions only with no explicit locale.
   */
  style: _NS("style"),

  /**
   * The superclass subclasses of which will be selected.
   *
   * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
   */
  category: _NS("category"),

  /**
   * Many fields prompt for information about a given property of the subject
   *
   * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
   */
  dependingOn: _NS("dependingOn"),

  /**
   * Max length
   *
   * This term has [3] labels (in languages [NoLocale, en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  maxLength: _NS("maxLength"),

  /**
   * part
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  part: _NS("part"),

  /**
   * Time field
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  TimeField: _NS("TimeField"),

  /**
   * A form which may be used to add more infromation to an
instance of this class which we know something about.  Anything from
adding just add one more fact, to adding a whole lot of information about a specific
facet of the thing.

   *
   * This term provides descriptions only with no explicit locale.
   */
  annotationForm: _NS("annotationForm"),

  /**
   * background color
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  backgroundColor: _NS("backgroundColor"),

  /**
   * URI or base64 representation of an image
   *
   * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
   */
  backgroundImage: _NS("backgroundImage"),

  /**
   * Base
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  base: _NS("base"),

  /**
   * color
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  color: _NS("color"),

  /**
   * Default error
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  defaultError: _NS("defaultError"),

  /**
   * A really simple way of enabling user interfaces to
            create new information about a class of things is to make a define of properties
            to be specified when a information about a new item
            ("New item" here means an item which the system
            does not have prvious information about yet,
            not an items which has just been created,
            like new friend as opposed to new baby)
   *
   * This term provides descriptions only with no explicit locale.
   */
  initialProperties: _NS("initialProperties"),

  /**
   * Maximum date offset
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  maxDateOffset: _NS("maxDateOffset"),

  /**
   * Maximum datetime offset
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  maxDatetimeOffset: _NS("maxDatetimeOffset"),

  /**
   * max
   *
   * This term has a label (in language [NoLocale]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  maxValue: _NS("maxValue"),

  /**
   * Minimum date offset
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  minDateOffset: _NS("minDateOffset"),

  /**
   * Minimum datetime offset
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  minDatetimeOffset: _NS("minDatetimeOffset"),

  /**
   * Min length
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  minLength: _NS("minLength"),

  /**
   * min
   *
   * This term has a label (in language [NoLocale]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   */
  minValue: _NS("minValue"),

  /**
   * Name
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  name: _NS("name"),

  /**
   * Old value
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  oldValue: _NS("oldValue"),

  /**
   * Parent property
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  parentProperty: _NS("parentProperty"),

  /**
   * Parts clone
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  partsClone: _NS("partsClone"),

  /**
   * Pattern
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  pattern: _NS("pattern"),

  /**
   * Reference
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  reference: _NS("reference"),

  /**
   * Required
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  required: _NS("required"),

  /**
   * Required error
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  requiredError: _NS("requiredError"),

  /**
   * The sequence in which this item is arranged with repect to other parts.
   *
   * This term provides descriptions only with no explicit locale.
   */
  seqeunce: _NS("seqeunce"),

  /**
   * A property which typically is used to sort
        members of a given class.
   *
   * This term provides descriptions only with no explicit locale.
   */
  sortBy: _NS("sortBy"),

  /**
   * When individuals or classes must be sorted, then
        if they are given different values of sortPriority a user agent can
        use this as a hint to how to present information.
   *
   * This term provides descriptions only with no explicit locale.
   */
  sortPriority: _NS("sortPriority"),

  /**
   * This  is a crude way of specifying a table-based
            view for objects of this class.
   *
   * This term provides descriptions only with no explicit locale.
   */
  tableProperties: _NS("tableProperties"),

  /**
   * Valid
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  valid: _NS("valid"),

  /**
   * Validation error
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  validationError: _NS("validationError"),

  /**
   * Value
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  value: _NS("value"),

  /**
   * Values
   *
   * This term has [2] labels (in languages [en, es]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/ns/ui
   */
  values: _NS("values"),

}

export default UI;
