// @inrupt/vocab-common-rdf/GeneratedVocab/VANN.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * A vocabulary for annotating vocabulary descriptions (VANN).
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://purl.org/vocab/vann/";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * A vocabulary for annotating vocabulary descriptions (VANN).
 */
const VANN = {
  PREFIX: "vann",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "vann": "http://purl.org/vocab/vann/" },
  NS: _NS,

  // *******************
  // All the Properties.
  // *******************

  /**
   * The preferred namespace prefix to ue when using terms from this vocabulary in an XML document.
   *
   * This term provides descriptions only in English.
   */
  preferredNamespacePrefix: _NS("preferredNamespacePrefix"),

  /**
   * The preferred namespace URI to use when using terms from this vocabulary in an XML document.
   *
   * This term provides descriptions only in English.
   */
  preferredNamespaceUri: _NS("preferredNamespaceUri"),

  /**
   * A reference to a resource that describes changes between this version of a vocabulary and the previous.
   *
   * This term provides descriptions only in English.
   */
  changes: _NS("changes"),

  /**
   * A reference to a resource that provides an example of how this resource can be used.
   *
   * This term provides descriptions only in English.
   */
  example: _NS("example"),

  /**
   * A group of related terms in a vocabulary.
   *
   * This term provides descriptions only in English.
   */
  termGroup: _NS("termGroup"),

  /**
   * A reference to a resource that provides information on how this resource is to be used.
   *
   * This term provides descriptions only in English.
   */
  usageNote: _NS("usageNote"),

}

export default VANN;
