// @inrupt/vocab-common-rdf/GeneratedVocab/VCARD.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * Ontology for vCard based on RFC6350
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://www.w3.org/2006/vcard/ns#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * Ontology for vCard based on RFC6350
 */
const VCARD = {
  PREFIX: "vcard",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "vcard": "http://www.w3.org/2006/vcard/ns#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * Acquaintance
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Acquaintance: _NS("Acquaintance"),

  /**
   * Used for relation type codes. The URI of the relation type code must be used as the value for the Relation Type.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  RelatedType: _NS("RelatedType"),

  /**
   * Agent
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Agent: _NS("Agent"),

  /**
   * This class is deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  BBS: _NS("BBS"),

  /**
   * Used for telephone type codes. The URI of the telephone type code must be used as the value for the Telephone Type.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  TelephoneType: _NS("TelephoneType"),

  /**
   * This class is deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Car: _NS("Car"),

  /**
   * Also called mobile telephone
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Cell: _NS("Cell"),

  /**
   * Child
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Child: _NS("Child"),

  /**
   * Colleague
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Colleague: _NS("Colleague"),

  /**
   * Contact
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Contact: _NS("Contact"),

  /**
   * Coresident
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Coresident: _NS("Coresident"),

  /**
   * Coworker
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Coworker: _NS("Coworker"),

  /**
   * Crush
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Crush: _NS("Crush"),

  /**
   * Date
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Date: _NS("Date"),

  /**
   * This class is deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Dom: _NS("Dom"),

  /**
   * Used for type codes. The URI of the type code must be used as the value for Type.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Type: _NS("Type"),

  /**
   * Emergency
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Emergency: _NS("Emergency"),

  /**
   * Fax
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Fax: _NS("Fax"),

  /**
   * Female
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Female: _NS("Female"),

  /**
   * Used for gender codes. The URI of the gender code must be used as the value for Gender.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Gender: _NS("Gender"),

  /**
   * Friend
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Friend: _NS("Friend"),

  /**
   * This implies that the property is related to an individual's personal life
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Home: _NS("Home"),

  /**
   * This class is deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  ISDN: _NS("ISDN"),

  /**
   * This class is deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Internet: _NS("Internet"),

  /**
   * This class is deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Intl: _NS("Intl"),

  /**
   * Kin
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Kin: _NS("Kin"),

  /**
   * This class is deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Label: _NS("Label"),

  /**
   * Male
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Male: _NS("Male"),

  /**
   * Me
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Me: _NS("Me"),

  /**
   * Met
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Met: _NS("Met"),

  /**
   * This class is deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Modem: _NS("Modem"),

  /**
   * This class is deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Msg: _NS("Msg"),

  /**
   * Muse
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Muse: _NS("Muse"),

  /**
   * Neighbor
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Neighbor: _NS("Neighbor"),

  /**
   * None
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  None: _NS("None"),

  /**
   * Other
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Other: _NS("Other"),

  /**
   * This class is deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  PCS: _NS("PCS"),

  /**
   * Pager
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Pager: _NS("Pager"),

  /**
   * This class is deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Parcel: _NS("Parcel"),

  /**
   * Parent
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Parent: _NS("Parent"),

  /**
   * This class is deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Postal: _NS("Postal"),

  /**
   * This class is deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Pref: _NS("Pref"),

  /**
   * Sibling
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Sibling: _NS("Sibling"),

  /**
   * Spouse
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Spouse: _NS("Spouse"),

  /**
   * Sweetheart
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Sweetheart: _NS("Sweetheart"),

  /**
   * This class is deprecated. Use the hasTelephone object property.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Tel: _NS("Tel"),

  /**
   * Also called sms telephone
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Text: _NS("Text"),

  /**
   * Text phone
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  TextPhone: _NS("TextPhone"),

  /**
   * Unknown
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Unknown: _NS("Unknown"),

  /**
   * Video
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Video: _NS("Video"),

  /**
   * Voice
   *
   * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Voice: _NS("Voice"),

  /**
   * This implies that the property is related to an individual's work place
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Work: _NS("Work"),

  /**
   * This class is deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  X400: _NS("X400"),

  /**
   * To specify the components of the delivery address for the  object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Address: _NS("Address"),

  /**
   * To specify the electronic mail address for communication with the object the vCard represents. Use the hasEmail object property.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Email: _NS("Email"),

  /**
   * Object representing a group of persons or entities.  A group object will usually contain hasMember properties to specify the members of the group.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Group: _NS("Group"),

  /**
   * The parent class for all objects
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Kind: _NS("Kind"),

  /**
   * An object representing a single person or entity
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Individual: _NS("Individual"),

  /**
   * An object representing a named geographical place
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Location: _NS("Location"),

  /**
   * An object representing an organization.  An organization is a single entity, and might represent a business or government, a department or division within a business or government, a club, an association, or the like.

   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Organization: _NS("Organization"),

  /**
   * The vCard class is  equivalent to the new Kind class, which is the parent for the four explicit types of vCards (Individual, Organization, Location, Group)
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  VCard: _NS("VCard"),

  /**
   * To specify the components of the name of the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  Name: _NS("Name"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * This object property has been mapped
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  adr: _NS("adr"),

  /**
   * To specify the components of the delivery address for the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasAddress: _NS("hasAddress"),

  /**
   * This object property has been deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  agent: _NS("agent"),

  /**
   * The date of marriage, or equivalent, of the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  anniversary: _NS("anniversary"),

  /**
   * To specify the birth date of the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  bday: _NS("bday"),

  /**
   * The category information about the object, also known as tags
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  category: _NS("category"),

  /**
   * This data property has been deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  class: _NS("class"),

  /**
   * This object property has been mapped
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  email: _NS("email"),

  /**
   * To specify the electronic mail address for communication with the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasEmail: _NS("hasEmail"),

  /**
   * This data property has been deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  extended_address: _NS("extended-address"),

  /**
   * This object property has been mapped
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  geo: _NS("geo"),

  /**
   * To specify information related to the global positioning of the object. May also be used as a property parameter.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasGeo: _NS("hasGeo"),

  /**
   * Used to support property parameters for the additional name data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasAdditionalName: _NS("hasAdditionalName"),

  /**
   * To specify the busy time associated with the object. (Was called FBURL in RFC6350)
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasCalendarBusy: _NS("hasCalendarBusy"),

  /**
   * To specify the calendar associated with the object. (Was called CALURI in RFC6350)
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasCalendarLink: _NS("hasCalendarLink"),

  /**
   * To specify the calendar user address to which a scheduling request be sent for the object. (Was called CALADRURI in RFC6350)
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasCalendarRequest: _NS("hasCalendarRequest"),

  /**
   * Used to support property parameters for the category data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasCategory: _NS("hasCategory"),

  /**
   * Used to support property parameters for the country name data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasCountryName: _NS("hasCountryName"),

  /**
   * Used to support property parameters for the formatted name data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasFN: _NS("hasFN"),

  /**
   * Used to support property parameters for the family name data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasFamilyName: _NS("hasFamilyName"),

  /**
   * To specify  the sex or gender identity of the object. URIs are recommended to enable interoperable sex and gender codes to be used.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasGender: _NS("hasGender"),

  /**
   * Used to support property parameters for the given name data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasGivenName: _NS("hasGivenName"),

  /**
   * Used to support property parameters for the honorific prefix data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasHonorificPrefix: _NS("hasHonorificPrefix"),

  /**
   * Used to support property parameters for the honorific suffix data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasHonorificSuffix: _NS("hasHonorificSuffix"),

  /**
   * To specify the instant messaging and presence protocol communications with the object. (Was called IMPP in RFC6350)
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasInstantMessage: _NS("hasInstantMessage"),

  /**
   * Used to support property parameters for the language data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasLanguage: _NS("hasLanguage"),

  /**
   * Used to support property parameters for the locality data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasLocality: _NS("hasLocality"),

  /**
   * Used to support property parameters for the nickname data property
   *
   * This term provides descriptions only in English.
   *
   * See also:
   *  - http://www.w3.org/2006/vcard/ns#nickname
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasNickname: _NS("hasNickname"),

  /**
   * The nick name associated with the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  nickname: _NS("nickname"),

  /**
   * Used to support property parameters for the note data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasNote: _NS("hasNote"),

  /**
   * Used to support property parameters for the organization name data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasOrganizationName: _NS("hasOrganizationName"),

  /**
   * Used to support property parameters for the organization unit name data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasOrganizationUnit: _NS("hasOrganizationUnit"),

  /**
   * Used to support property parameters for the postal code data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasPostalCode: _NS("hasPostalCode"),

  /**
   * Used to support property parameters for the region data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasRegion: _NS("hasRegion"),

  /**
   * To specify a relationship between another entity and the entity represented by this object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasRelated: _NS("hasRelated"),

  /**
   * Used to support property parameters for the role data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasRole: _NS("hasRole"),

  /**
   * To identify the source of directory information of the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasSource: _NS("hasSource"),

  /**
   * Used to support property parameters for the street address data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasStreetAddress: _NS("hasStreetAddress"),

  /**
   * Used to support property parameters for the title data property
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasTitle: _NS("hasTitle"),

  /**
   * To specify a value that represents a globally unique identifier corresponding to the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasUID: _NS("hasUID"),

  /**
   * Used to indicate the resource value of an object property that requires property parameters
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasValue: _NS("hasValue"),

  /**
   * This data property has been deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  label: _NS("label"),

  /**
   * To specify the language that may be used for contacting the object. May also be used as a property parameter.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  language: _NS("language"),

  /**
   * This data property has been deprecated. See hasGeo
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  latitude: _NS("latitude"),

  /**
   * This data property has been deprecated. See hasGeo
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  longitude: _NS("longitude"),

  /**
   * This data property has been deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  mailer: _NS("mailer"),

  /**
   * A note associated with the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  note: _NS("note"),

  /**
   * This object property has been mapped. Use the organization-name data property.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  org: _NS("org"),

  /**
   * To specify the organizational name associated with the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  organization_name: _NS("organization-name"),

  /**
   * To specify the organizational unit name associated with the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  organization_unit: _NS("organization-unit"),

  /**
   * This data property has been deprecated
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  post_office_box: _NS("post-office-box"),

  /**
   * To specify the identifier for the product that created the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  prodid: _NS("prodid"),

  /**
   * To specify revision information about the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  rev: _NS("rev"),

  /**
   * To specify the function or part played in a particular situation by the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  role: _NS("role"),

  /**
   * To specify the string to be used for national-language-specific sorting. Used as a property parameter only.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  sort_string: _NS("sort-string"),

  /**
   * To specify the position or job of the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  title: _NS("title"),

  /**
   * To indicate time zone information that is specific to the object. May also be used as a property parameter.
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  tz: _NS("tz"),

  /**
   * Used to indicate the literal value of a data property that requires property parameters
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  value: _NS("value"),

  /**
   * The country name associated with the address of the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  country_name: _NS("country-name"),

  /**
   * The locality (e.g. city or town) associated with the address of the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  locality: _NS("locality"),

  /**
   * The postal code associated with the address of the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  postal_code: _NS("postal-code"),

  /**
   * The region (e.g. state or province) associated with the address of the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  region: _NS("region"),

  /**
   * The street address associated with the address of the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  street_address: _NS("street-address"),

  /**
   * The formatted text corresponding to the name of the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  fn: _NS("fn"),

  /**
   * To include a member in the group this object represents. (This property can only be used by Group individuals)
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasMember: _NS("hasMember"),

  /**
   * The additional name associated with the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  additional_name: _NS("additional-name"),

  /**
   * The family name associated with the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  family_name: _NS("family-name"),

  /**
   * The given name associated with the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  given_name: _NS("given-name"),

  /**
   * The honorific prefix of the name associated with the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  honorific_prefix: _NS("honorific-prefix"),

  /**
   * The honorific suffix of the name associated with the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  honorific_suffix: _NS("honorific-suffix"),

  /**
   * To specify a public key or authentication certificate associated with the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasKey: _NS("hasKey"),

  /**
   * This object property has been mapped
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  key: _NS("key"),

  /**
   * To specify a graphic image of a logo associated with the object 
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasLogo: _NS("hasLogo"),

  /**
   * This object property has been mapped
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  logo: _NS("logo"),

  /**
   * To specify the components of the name of the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasName: _NS("hasName"),

  /**
   * This object property has been mapped
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  n: _NS("n"),

  /**
   * To specify an image or photograph information that annotates some aspect of the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasPhoto: _NS("hasPhoto"),

  /**
   * This object property has been mapped
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  photo: _NS("photo"),

  /**
   * To specify a digital sound content information that annotates some aspect of the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasSound: _NS("hasSound"),

  /**
   * This object property has been mapped
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  sound: _NS("sound"),

  /**
   * To specify the telephone number for telephony communication with the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasTelephone: _NS("hasTelephone"),

  /**
   * This object property has been mapped
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  tel: _NS("tel"),

  /**
   * To specify a uniform resource locator associated with the object
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  hasURL: _NS("hasURL"),

  /**
   * This object property has been mapped
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2006/vcard/ns
   */
  url: _NS("url"),

}

export default VCARD;
