// @inrupt/vocab-common-rdf/GeneratedVocab/XSD.ts
/**
 * MIT License
 * 
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-common-rdf], version: [1.0.5]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * A vocabulary for describing XML Schema datatypes.
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://www.w3.org/2001/XMLSchema#";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * A vocabulary for describing XML Schema datatypes.
 */
const XSD = {
  PREFIX: "xsd",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "xsd": "http://www.w3.org/2001/XMLSchema#" },
  NS: _NS,

  // *******************
  // All the Properties.
  // *******************

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  changes: _NS("changes"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  string: _NS("string"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  boolean: _NS("boolean"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  float: _NS("float"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  double: _NS("double"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  decimal: _NS("decimal"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  dateTime: _NS("dateTime"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  duration: _NS("duration"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  hexBinary: _NS("hexBinary"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  base64Binary: _NS("base64Binary"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  anyURI: _NS("anyURI"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  ID: _NS("ID"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  IDREF: _NS("IDREF"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  ENTITY: _NS("ENTITY"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  NOTATION: _NS("NOTATION"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  normalizedString: _NS("normalizedString"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  token: _NS("token"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  language: _NS("language"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  IDREFS: _NS("IDREFS"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  ENTITIES: _NS("ENTITIES"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  NMTOKEN: _NS("NMTOKEN"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  NMTOKENS: _NS("NMTOKENS"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  Name: _NS("Name"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  QName: _NS("QName"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  NCName: _NS("NCName"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  integer: _NS("integer"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  nonNegativeInteger: _NS("nonNegativeInteger"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  positiveInteger: _NS("positiveInteger"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  nonPositiveInteger: _NS("nonPositiveInteger"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  negativeInteger: _NS("negativeInteger"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  byte: _NS("byte"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  int: _NS("int"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  long: _NS("long"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  short: _NS("short"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  unsignedByte: _NS("unsignedByte"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  unsignedInt: _NS("unsignedInt"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  unsignedLong: _NS("unsignedLong"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  unsignedShort: _NS("unsignedShort"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  date: _NS("date"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  time: _NS("time"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  gYearMonth: _NS("gYearMonth"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  gYear: _NS("gYear"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  gMonthDay: _NS("gMonthDay"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  gDay: _NS("gDay"),

  /**
   * .
   *
   * This term provides descriptions only in English.
   *
   * Defined by the vocabulary: http://www.w3.org/2001/XMLSchema#
   */
  gMonth: _NS("gMonth"),

}

export default XSD;
