// generated-vocab-pwv/GeneratedVocab/PWV.js
/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [3.0.2]
 * as part of artifact: [generated-vocab-pwv], version: [0.0.1]
 * on 'Saturday, September 30, 2023 3:31 PM'.
 *
 * Vocabulary built from input: [https://jeffreywallphd.github.io/vocab/pwv.ttl].
 * The generator detected the following terms in the source vocabulary:
 *  - Classes: [13]
 *  - Properties: [1]
 *  - Literals: [0]
 *  - Constant IRIs: [0]
 *  - Constant strings: [0]
 *
 * Vocabulary for personal wellness goal setting and activity tracking
 *
 * Inrupt Best Practice Compliance Report:
 *  - PASS - namespace IRI provided explicitly. [inrupt_bp:namespace_IriProvidedExplicitly]
 *    - Detail - namespace IRI provided via VANN:preferredNamespaceUri. [inrupt_bp:namespace_IriProvidedExplicitly_definedUsingVann]
 * 
 *  - PASS - namespace prefix provided explicitly. [inrupt_bp:namespace_PrefixProvidedExplicitly]
 *    - Detail - namespace prefix provided via SHACL:declare. [inrupt_bp:namespace_PrefixProvidedExplicitly_definedUsingShacl]
 *    - Detail - multiple namespace prefixes provided via SHACL:declare and SHACL:prefix values of [XXX,YYY]. [inrupt_bp:namespace_PrefixProvidedExplicitly_definedUsingShacl_multipleValues]
 * 
 *  - FAIL - Namespace IRI different from vocabulary IRI. [inrupt_bp:namespace_IriDifferentFromVocabularyIri]
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "https://jeffreywallphd.github.io/vocab/pwv.ttl#";
function _NS (localName) {
  return (_NAMESPACE + localName);
}

/**
 * Vocabulary for personal wellness goal setting and activity tracking
 *
 * Namespace IRI: [https://jeffreywallphd.github.io/vocab/pwv.ttl#]
 */
const PWV = {
  PREFIX: "pwv",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "pwv": "https://jeffreywallphd.github.io/vocab/pwv.ttl#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * An abstract activity performed by a person
   *
   * This term provides descriptions only with no explicit locale.
   */
  Activity: _NS("Activity"),

  /**
   * An physical activity performed by a person
   *
   * This term provides descriptions only with no explicit locale.
   */
  PhysicalActivity: _NS("PhysicalActivity"),

  /**
   * A nutritional activity performed by a person
   *
   * This term provides descriptions only with no explicit locale.
   */
  NutritionalActivity: _NS("NutritionalActivity"),

  /**
   * A mental activity performed by a person
   *
   * This term provides descriptions only with no explicit locale.
   */
  MentalActivity: _NS("MentalActivity"),

  /**
   * A social activity performed by a person
   *
   * This term provides descriptions only with no explicit locale.
   */
  SocialActivity: _NS("SocialActivity"),

  /**
   * A sleep activity performed by a person
   *
   * This term provides descriptions only with no explicit locale.
   */
  SleepActivity: _NS("SleepActivity"),

  /**
   * An intellectual activity performed by a person
   *
   * This term provides descriptions only with no explicit locale.
   */
  IntellectualActivity: _NS("IntellectualActivity"),

  /**
   * A spiritual activity performed by a person
   *
   * This term provides descriptions only with no explicit locale.
   */
  SpiritualActivity: _NS("SpiritualActivity"),

  /**
   * A career activity performed by a person
   *
   * This term provides descriptions only with no explicit locale.
   */
  CareerActivity: _NS("CareerActivity"),

  /**
   * A financial activity performed by a person
   *
   * This term provides descriptions only with no explicit locale.
   */
  FinancialActivity: _NS("FinancialActivity"),

  /**
   * A measurement of an activity, person, or goal
   *
   * This term provides descriptions only with no explicit locale.
   */
  Measurement: _NS("Measurement"),

  /**
   * An imperial measurement of an activity, person, or goal
   *
   * This term provides descriptions only with no explicit locale.
   */
  ImperialMeasurement: _NS("ImperialMeasurement"),

  /**
   * A metric measurement of an activity, person, or goal
   *
   * This term provides descriptions only with no explicit locale.
   */
  MetricMeasurement: _NS("MetricMeasurement"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * The act of measuring some activity, person, or goal using a measurement
   *
   * This term provides descriptions only with no explicit locale.
   */
  measuredBy: _NS("measuredBy"),

}

module.exports = PWV;
