// @inrupt/vocab-solid/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/**
 * MIT License
 *
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * [Generator provided] - Web Access Control Vocabulary
 */
// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
var _NAMESPACE$a = "http://www.w3.org/ns/auth/acl#";
function _NS$a(localName) {
    return (_NAMESPACE$a + localName);
}
/**
 * [Generator provided] - Web Access Control Vocabulary
 */
var ACL = {
    PREFIX: "acl",
    NAMESPACE: _NAMESPACE$a,
    PREFIX_AND_NAMESPACE: { "acl": "http://www.w3.org/ns/auth/acl#" },
    NS: _NS$a,
    // *****************
    // All the Classes.
    // *****************
    /**
     * Any kind of access to a resource. Don't use this, use R W and RW
     *
     * The term has a description only in English, but has a mismatch between its labels and comments, with [0] labels, but [1] comment in the language [NoLocale].
     */
    Access: _NS$a("Access"),
    /**
     * Append accesses are specific write access which only add information, and do not remove information.
      For text files, for example, append access allows bytes to be added onto the end of the file.
      For RDF graphs, Append access allows adds triples to the graph but does not remove any.
      Append access is useful for dropbox functionality.
      Dropbox can be used for link notification, which the information added is a notification
      that a some link has been made elsewhere relevant to the given resource.
      
     *
     * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
     */
    Append: _NS$a("Append"),
    /**
     * write
     *
     * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
     */
    Write: _NS$a("Write"),
    /**
     * A class of agents who have been authenticated.
  In other words, anyone can access this resource, but not anonymously.
  The social expectation is that the authentication process will provide an
  identify and a name, or pseudonym.
  (A new ID should not be minted for every access: the intent is that the user
  is able to continue to use the ID for continues interactions with peers,
  and for example to develop a reputation)
  
     *
     * This term provides descriptions only with no explicit locale.
     */
    AuthenticatedAgent: _NS$a("AuthenticatedAgent"),
    /**
     * An element of access control,
      allowing agent to agents access of some kind to resources or classes of resources
     *
     * This term provides descriptions only with no explicit locale.
     */
    Authorization: _NS$a("Authorization"),
    /**
     * Allows read/write access to the ACL for the resource(s)
     *
     * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
     */
    Control: _NS$a("Control"),
    /**
     * An Origin is basically a web site
          (Note WITHOUT the trailing slash after the domain name and port in its URI)
          and is the basis for controlling access to data by web apps
          in the Same Origin Model of web security.
          All scripts from the same origin are given the same right.
     *
     * This term provides descriptions only with no explicit locale.
     *
     * See also:
     *  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin
     */
    Origin: _NS$a("Origin"),
    /**
     * The class of read operations
     *
     * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
     */
    Read: _NS$a("Read"),
    // *******************
    // All the Properties.
    // *******************
    /**
     * The Access Control file for this information resource.
          This may of course be a virtual resource implemented by the access control system.
          Note also HTTP's header  Link:  foo.meta ;rel=meta can be used for this.
     *
     * This term provides descriptions only with no explicit locale.
     */
    accessControl: _NS$a("accessControl"),
    /**
     * The information resource to which access is being granted.
     *
     * This term provides descriptions only with no explicit locale.
     */
    accessTo: _NS$a("accessTo"),
    /**
     * A class of information resources to which access is being granted.
     *
     * This term provides descriptions only with no explicit locale.
     */
    accessToClass: _NS$a("accessToClass"),
    /**
     * A person or social entity to being given the right
     *
     * This term provides descriptions only with no explicit locale.
     */
    agent: _NS$a("agent"),
    /**
     * A class of persons or social entities to being given the right
     *
     * This term provides descriptions only with no explicit locale.
     */
    agentClass: _NS$a("agentClass"),
    /**
     * A group of persons or social entities to being given the right.
            The right is given to any entity which is a vcard:member of the group,
            as defined by the document received when the Group is dereferenced.
     *
     * This term provides descriptions only with no explicit locale.
     */
    agentGroup: _NS$a("agentGroup"),
    /**
     * If a resource has no ACL file (it is 404),
          then access to the resource if given by the ACL of the immediately
          containing directory, or failing that (404) the ACL of the recursively next
          containing directory which has an ACL file.
          Within that ACL file,
          any Authentication which has that directory as its acl:default applies to the
          resource. (The highest directory must have an ACL file.)
  
     *
     * This term provides descriptions only with no explicit locale.
     */
    default: _NS$a("default"),
    /**
     * THIS IS OBSOLETE AS OF 2017-08-01.   See 'default'.
          Was: A directory for which this authorization is used for new files in the directory.
     *
     * This term provides descriptions only with no explicit locale.
     */
    defaultForNew: _NS$a("defaultForNew"),
    /**
     * Delegates a person or another agent to act on behalf of the agent.
      For example, Alice delegates Bob to act on behalf of Alice for ACL purposes.
     *
     * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
     */
    delegates: _NS$a("delegates"),
    /**
     * A mode of access such as read or write.
     *
     * This term provides descriptions only with no explicit locale.
     */
    mode: _NS$a("mode"),
    /**
     * A web application, identified by its Origin, such as
          <https://scripts.example.com>, being given the right.
          When a user of the web application at a certain origin accesses the server,
          then the browser sets the Origin: header to warn that a possibly untrusted webapp
          is being used.
          Then, BOTH the user AND the origin must have the required access.
     *
     * This term provides descriptions only with no explicit locale.
     *
     * See also:
     *  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin
     */
    origin: _NS$a("origin"),
    /**
     * The person or other agent which owns this.
      For example, the owner of a file in a filesystem.
      There is a sense of right to control.   Typically defaults to the agent who craeted
      something but can be changed.
     *
     * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
     */
    owner: _NS$a("owner"),
};

/**
 * MIT License
 *
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * A vocabulary to structure policy-based access controls for the Solid ecosystem.
 */
// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
var _NAMESPACE$9 = "http://www.w3.org/ns/solid/acp#";
function _NS$9(localName) {
    return (_NAMESPACE$9 + localName);
}
/**
 * A vocabulary to structure policy-based access controls for the Solid ecosystem.
 */
var ACP = {
    PREFIX: "acp",
    NAMESPACE: _NAMESPACE$9,
    PREFIX_AND_NAMESPACE: { "acp": "http://www.w3.org/ns/solid/acp#" },
    NS: _NS$9,
    // *****************
    // All the Classes.
    // *****************
    /**
     * Access Control statements associate an AccessControlResource with specific Policy definitions.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    AccessControl: _NS$9("AccessControl"),
    /**
     * An RDF resource that includes AccessControl statements.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    AccessControlResource: _NS$9("AccessControlResource"),
    /**
     * A Policy defines a collection of access modes along with the conditions under which the policy is applied.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    Policy: _NS$9("Policy"),
    /**
     * A Rule defines which agent(s), group(s) and client application(s) match a Policy filter.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    Rule: _NS$9("Rule"),
    /**
     * An abstract access mode. This class should not be used directly. Instead, please use Read, Write and Append.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    AccessMode: _NS$9("AccessMode"),
    /**
     * An access mode indicating Read access
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    Read: _NS$9("Read"),
    /**
     * An access mode indicating Write access.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    Write: _NS$9("Write"),
    /**
     * An access mode indicating Append access. Append is a more limited form of Write in which data may only be added but not removed.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    Append: _NS$9("Append"),
    // *******************
    // All the Properties.
    // *******************
    /**
     * An IRI representing an authenticated agent.
   Please note that in a decentralized authentication ecosystem, an authenticated agent may be ephemeral
   and, from a security standpoint, may not be significantly different than an anonymous agent.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    AuthenticatedAgent: _NS$9("AuthenticatedAgent"),
    /**
     * An IRI representing the creator of a resource.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    CreatorAgent: _NS$9("CreatorAgent"),
    /**
     * An IRI representing a public or anonymous agent.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    PublicAgent: _NS$9("PublicAgent"),
    /**
     * The object of this relationship is the access control resource associated with the subject resource.
   This relationship is intended to be used as the rel type of a Link, via HTTP Link Headers rfc5988 for Linked Data resources.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    accessControl: _NS$9("accessControl"),
    /**
     * The access property identifies the access policies that apply to the ACR itself.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    access: _NS$9("access"),
    /**
     * The accessLocked property identifies the access policies that apply to the ACR itself.
   These policies may not be removed except by an agent with write access to the root ACR.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    accessLocked: _NS$9("accessLocked"),
    /**
     * The accessProtected property identifies the access policies that apply to the ACR itself.
   These policies may not be removed except by an agent with write access to the ACR where the policy was initially applied.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    accessProtected: _NS$9("accessProtected"),
    /**
     * The accessMembers property identifies the access policies that apply to the ACR itself, recursively.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    accessMembers: _NS$9("accessMembers"),
    /**
     * The accessMembersLocked property identifies the access policies that apply to the ACR itself, recursively.
   These policies may not be removed except by an agent with write access to the root ACR.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    accessMembersLocked: _NS$9("accessMembersLocked"),
    /**
     * The accessMembersProtected property identifies the access policies that apply to the ACR itself, recursively.
   These policies may not be removed except by an agent with write access to the ACR where the policy was initially applied.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    accessMembersProtected: _NS$9("accessMembersProtected"),
    /**
     * The apply property identifies the access policies to apply to a resource.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    apply: _NS$9("apply"),
    /**
     * The applyProtected property identifies the access policies to apply to a resource's members recursively.
   These policies may not be removed except by an agent with write access to the ACR where the policy was initially applied.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    applyProtected: _NS$9("applyProtected"),
    /**
     * The applyLocked property identifies the access policies to apply to a resource's members recursively.
   These policies may not be removed except by an agent with write access to the root ACR.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    applyLocked: _NS$9("applyLocked"),
    /**
     * The applyMembers property identifies the access policies to apply to a resource's members recursively.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    applyMembers: _NS$9("applyMembers"),
    /**
     * The applyMembersProtected property identifies the access policies to apply to a resource's members recursively.
   These policies may not be removed except by an agent with write access to the ACR where the policy was initially applied.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    applyMembersProtected: _NS$9("applyMembersProtected"),
    /**
     * The applyMembersLocked property identifies the access policies to apply to a resource's members recursively.
   These policies may not be removed except by an agent with write access to the root ACR.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    applyMembersLocked: _NS$9("applyMembersLocked"),
    /**
     * The allOf property identifies a collection of Matchers, each of which must match in order for the Policy to take effect.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    allOf: _NS$9("allOf"),
    /**
     * The anyOf property identifies a collection of Matchers, one of which must match in order for the Policy to take effect.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    anyOf: _NS$9("anyOf"),
    /**
     * The noneOf property identifies a collection of Matchers, none of which must match in order for the Policy to take effect.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    noneOf: _NS$9("noneOf"),
    /**
     * The allow property identifies the access modes a Policy allows.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    allow: _NS$9("allow"),
    /**
     * The deny property identifies the access modes a Policy denies.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    deny: _NS$9("deny"),
    /**
     * The agent property identifies the agent WebID(s) to which a Matcher matches.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    agent: _NS$9("agent"),
    /**
     * The client property identifies the client WebID(s) to which a Matcher matches.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    client: _NS$9("client"),
    /**
     * The group property identifies the vCard group(s) to which a Matcher matches.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/acp#
     */
    group: _NS$9("group"),
};

/**
 * MIT License
 *
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * The OpenID Connect vocabulary used by the Solid-OIDC authentication specification.
 */
// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
var _NAMESPACE$8 = "http://www.w3.org/ns/solid/oidc#";
function _NS$8(localName) {
    return (_NAMESPACE$8 + localName);
}
/**
 * The OpenID Connect vocabulary used by the Solid-OIDC authentication specification.
 */
var OIDC = {
    PREFIX: "oidc",
    NAMESPACE: _NAMESPACE$8,
    PREFIX_AND_NAMESPACE: { "oidc": "http://www.w3.org/ns/solid/oidc#" },
    NS: _NS$8,
    // *******************
    // All the Properties.
    // *******************
    /**
     * TODO
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
     */
    application_type: _NS$8("application_type"),
    /**
     * A name for this client that can be presented to an End-User
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
     */
    client_name: _NS$8("client_name"),
    /**
     * A URI for this client's home page
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
     */
    client_uri: _NS$8("client_uri"),
    /**
     * A collection of URIs where individuals or entities responsible for this client can be contacted.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
     */
    contacts: _NS$8("contacts"),
    /**
     * TODO
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
     */
    default_max_age: _NS$8("default_max_age"),
    /**
     * A collection of OAuth 2.0 Grant Types that the client will restrict itself to using.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
     */
    grant_types: _NS$8("grant_types"),
    /**
     * A URI for this client's logo
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
     */
    logo_uri: _NS$8("logo_uri"),
    /**
     * A URI for this client's policy document
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
     */
    policy_uri: _NS$8("policy_uri"),
    /**
     * A collection of registered URIs used by the client for redirection
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
     */
    redirect_uris: _NS$8("redirect_uris"),
    /**
     * TODO
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
     */
    require_auth_time: _NS$8("require_auth_time"),
    /**
     * A collection of OAuth 2.0 Response Types that the client will restrict itself to using.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
     */
    response_types: _NS$8("response_types"),
    /**
     * An OAuth2 scope for the requested access token.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
     */
    scope: _NS$8("scope"),
    /**
     * The requested client authentication method for the Token Endpoint
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
     */
    token_endpoint_auth_method: _NS$8("token_endpoint_auth_method"),
    /**
     * A URI for this client's terms of service document
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/oidc#
     */
    tos_uri: _NS$8("tos_uri"),
};

/**
 * MIT License
 *
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * The Solid Terms vocabulary defines terms referenced in Solid specifications.
 */
// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
var _NAMESPACE$7 = "http://www.w3.org/ns/solid/terms#";
function _NS$7(localName) {
    return (_NAMESPACE$7 + localName);
}
/**
 * The Solid Terms vocabulary defines terms referenced in Solid specifications.
 */
var SOLID = {
    PREFIX: "solid",
    NAMESPACE: _NAMESPACE$7,
    PREFIX_AND_NAMESPACE: { "solid": "http://www.w3.org/ns/solid/terms#" },
    NS: _NS$7,
    // *****************
    // All the Classes.
    // *****************
    /**
     * A Solid account.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    Account: _NS$7("Account"),
    /**
     * A resource containing notifications.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    Inbox: _NS$7("Inbox"),
    /**
     * Listed Type Index is a registry of resources that are publicly discoverable by outside users and applications.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    ListedDocument: _NS$7("ListedDocument"),
    /**
     * A notification resource.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    Notification: _NS$7("Notification"),
    /**
     * A patch expresses conditional modifications to a resource that has an RDF-based representation.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    Patch: _NS$7("Patch"),
    /**
     * A resource containing time ordered items and sub-containers.  Sub-containers may be desirable in file based systems to split the timeline into logical components e.g. /yyyy-mm-dd/ as used in ISO 8061.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    Timeline: _NS$7("Timeline"),
    /**
     * A index of type registries for resources. Applications can register the RDF type they use and list them in the index resource.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    TypeIndex: _NS$7("TypeIndex"),
    /**
     * The registered types that map a RDF classes/types to their locations using either `instance` or `instanceContainer` property.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    TypeRegistration: _NS$7("TypeRegistration"),
    /**
     * Unlisted Type Index is a registry of resources that are private to the user and their apps, for types that are not publicly discoverable.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    UnlistedDocument: _NS$7("UnlistedDocument"),
    /**
     * The class of WebIDs - where a WebID is an HTTP URI which refers to an Agent (Person, Organization, Device, etc.).
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/solid#
     */
    WebId: _NS$7("WebId"),
    // *******************
    // All the Properties.
    // *******************
    /**
     * A solid account belonging to an Agent.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    account: _NS$7("account"),
    /**
     * The triple patterns this patch removes from the document.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    deletes: _NS$7("deletes"),
    /**
     * A class that is used to map an listed or unlisted type index.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    forClass: _NS$7("forClass"),
    /**
     * Deprecated pointer to a Linked Data Notifications inbox; please use http://www.w3.org/ns/ldp#inbox instead.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    inbox: _NS$7("inbox"),
    /**
     * The triple patterns this patch adds to the document.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    inserts: _NS$7("inserts"),
    /**
     * Maps a type to an individual resource, typically an index or a directory listing resource.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    instance: _NS$7("instance"),
    /**
     * Maps a type to a container which the client would have to list to get the instances of that type.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    instanceContainer: _NS$7("instanceContainer"),
    /**
     * The login URI of a given server.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    loginEndpoint: _NS$7("loginEndpoint"),
    /**
     * The logout URI of a given server.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    logoutEndpoint: _NS$7("logoutEndpoint"),
    /**
     * Notification resource for an inbox.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    notification: _NS$7("notification"),
    /**
     * The preferred OpenID Connect issuer URI for a given WebID.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    oidcIssuer: _NS$7("oidcIssuer"),
    /**
     * The document to which this patch applies.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    patches: _NS$7("patches"),
    /**
     * Points to an unlisted type index resource.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    privateTypeIndex: _NS$7("privateTypeIndex"),
    /**
     * Points to a listed type index resource.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    publicTypeIndex: _NS$7("publicTypeIndex"),
    /**
     * Points to a TypeIndex resource.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    typeIndex: _NS$7("typeIndex"),
    /**
     * Indicates if a message has been read or not. This property should have a boolean datatype.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    read: _NS$7("read"),
    /**
     * The quota of non-volatile memory that is available for the account (in bytes)
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    storageQuota: _NS$7("storageQuota"),
    /**
     * The amount of non-volatile memory that the account have used (in bytes)
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    storageUsage: _NS$7("storageUsage"),
    /**
     * Timeline for a given resource.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    timeline: _NS$7("timeline"),
    /**
     * The conditions the document and the inserted and deleted triple patterns need to satisfy in order for the patch to be applied.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: http://www.w3.org/ns/solid/terms#
     */
    where: _NS$7("where"),
    /**
     * A WebID is an HTTP URI which refers to an Agent (Person, Organization, Device, etc.).
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/solid#
     */
    webId: _NS$7("webId"),
    /**
     * A person or social entity that is considered to have control, rights, and responsibilities over a data storage.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/solid#
     */
    owner: _NS$7("owner"),
    /**
     * The IRI of the root resource of a Solid Pod.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/solid#
     */
    storageRoot: _NS$7("storageRoot"),
};

/**
 * MIT License
 *
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * The Solid vocabulary providing terms (e.g., classes, properties and text strings (e.g.
 informative labels or error messages) used by the Solid ACL system.
 */
// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
var _NAMESPACE$6 = "https://inrupt.com/vocab/solid_meta_acl#";
function _NS$6(localName) {
    return (_NAMESPACE$6 + localName);
}
/**
 * The Solid vocabulary providing terms (e.g., classes, properties and text strings (e.g.
 informative labels or error messages) used by the Solid ACL system.
 */
var SOLID_META_ACL = {
    PREFIX: "solid_meta_acl",
    NAMESPACE: _NAMESPACE$6,
    PREFIX_AND_NAMESPACE: { "solid_meta_acl": "https://inrupt.com/vocab/solid_meta_acl#" },
    NS: _NS$6,
    // *******************
    // All the Properties.
    // *******************
    /**
     * The root subject for ACL metadata.
   The idea for subject roots is that we could build up an entire graph of
   metadata (i.e., as processing progresses along a chain of processors, with each
   processor enriching this metadata with their specific metadata), and this
   term defines the root term of that graph.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: https://inrupt.com/vocab/solid_meta_acl#
     */
    subjectRoot: _NS$6("subjectRoot"),
    /**
     * The URI that contains the ACL
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: https://inrupt.com/vocab/solid_meta_acl#
     */
    aclUri: _NS$6("aclUri"),
};

/**
 * MIT License
 *
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * The Solid vocabulary providing terms (e.g., classes, properties and text strings (e.g.
    informative labels or error messages) used by Solid clients.
 */
// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
var _NAMESPACE$5 = "https://inrupt.com/vocab/solid_meta_client#";
function _NS$5(localName) {
    return (_NAMESPACE$5 + localName);
}
/**
 * The Solid vocabulary providing terms (e.g., classes, properties and text strings (e.g.
    informative labels or error messages) used by Solid clients.
 */
var SOLID_META_CLIENT = {
    PREFIX: "solid_meta_client",
    NAMESPACE: _NAMESPACE$5,
    PREFIX_AND_NAMESPACE: { "solid_meta_client": "https://inrupt.com/vocab/solid_meta_client#" },
    NS: _NS$5,
    // *******************
    // All the Properties.
    // *******************
    /**
     * The root subject for client-generated metadata.
   The idea for subject roots is that we could build up an entire graph of
   metadata (i.e., as processing progresses along a chain of processors, with each
   processor enriching this metadata with their specific metadata), and this
   term defines the root term of that graph.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: https://inrupt.com/vocab/solid_meta_client#
     */
    subjectRoot: _NS$5("subjectRoot"),
};

/**
 * MIT License
 *
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * The Solid vocabulary providing terms (e.g., classes, properties and text strings (e.g.
 informative labels or error messages) used by Solid servers at server creation time.
 */
// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
var _NAMESPACE$4 = "https://inrupt.com/vocab/solid_meta_server_creational#";
function _NS$4(localName) {
    return (_NAMESPACE$4 + localName);
}
/**
 * The Solid vocabulary providing terms (e.g., classes, properties and text strings (e.g.
 informative labels or error messages) used by Solid servers at server creation time.
 */
var SOLID_META_SERVER_CREATIONAL = {
    PREFIX: "solid_meta_server_creational",
    NAMESPACE: _NAMESPACE$4,
    PREFIX_AND_NAMESPACE: { "solid_meta_server_creational": "https://inrupt.com/vocab/solid_meta_server_creational#" },
    NS: _NS$4,
    // *******************
    // All the Properties.
    // *******************
    /**
     * The root subject for creation-time server-generated metadata.
   The idea for subject roots is that we could build up an entire graph of
   metadata (i.e., as processing progresses along a chain of processors, with each
   processor enriching this metadata with their specific metadata), and this
   term defines the root term of that graph.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: https://inrupt.com/vocab/solid_meta_server_creational#
     */
    subjectRoot: _NS$4("subjectRoot"),
    /**
     * Host name of the server processing this request.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: https://inrupt.com/vocab/solid_meta_server_creational#
     */
    hostName: _NS$4("hostName"),
    /**
     * Host IP address of the server processing this request.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: https://inrupt.com/vocab/solid_meta_server_creational#
     */
    hostIpAddress: _NS$4("hostIpAddress"),
    /**
     * Timestamp this server was instantiated.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: https://inrupt.com/vocab/solid_meta_server_creational#
     */
    createdTimestamp: _NS$4("createdTimestamp"),
};

/**
 * MIT License
 *
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * The Solid vocabulary providing terms (e.g., classes, properties and text strings (e.g.
 informative labels or error messages) used by Solid servers at server creation time.
 */
// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
var _NAMESPACE$3 = "https://inrupt.com/vocab/solid_meta_server_runtime#";
function _NS$3(localName) {
    return (_NAMESPACE$3 + localName);
}
/**
 * The Solid vocabulary providing terms (e.g., classes, properties and text strings (e.g.
 informative labels or error messages) used by Solid servers at server creation time.
 */
var SOLID_META_SERVER_RUNTIME = {
    PREFIX: "solid_meta_server_runtime",
    NAMESPACE: _NAMESPACE$3,
    PREFIX_AND_NAMESPACE: { "solid_meta_server_runtime": "https://inrupt.com/vocab/solid_meta_server_runtime#" },
    NS: _NS$3,
    // *******************
    // All the Properties.
    // *******************
    /**
     * The root subject for runtime server-generated metadata.
   The idea for subject roots is that we could build up an entire graph of
   metadata (i.e., as processing progresses along a chain of processors, with each
   processor enriching this metadata with their specific metadata), and this
   term defines the root term of that graph.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: https://inrupt.com/vocab/solid_meta_server_runtime#
     */
    subjectRoot: _NS$3("subjectRoot"),
    /**
     * Timestamp this request was received at this server.
     *
     * This term provides descriptions only in English.
     *
     * Defined by the vocabulary: https://inrupt.com/vocab/solid_meta_server_runtime#
     */
    requestReceivedTimestamp: _NS$3("requestReceivedTimestamp"),
};

/**
 * MIT License
 *
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * The Solid Service Description vocabulary describes the services provided by a Solid Server.
 */
// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
var _NAMESPACE$2 = "http://www.w3.org/ns/solid/service-description#";
function _NS$2(localName) {
    return (_NAMESPACE$2 + localName);
}
/**
 * The Solid Service Description vocabulary describes the services provided by a Solid Server.
 */
var SOLID_SD = {
    PREFIX: "solid_sd",
    NAMESPACE: _NAMESPACE$2,
    PREFIX_AND_NAMESPACE: { "solid_sd": "http://www.w3.org/ns/solid/service-description#" },
    NS: _NS$2,
    // *****************
    // All the Classes.
    // *****************
    /**
     * A Solid Service is some function provided by a Solid Server,
   e.g., to search a Pod, send a notification to an Inbox, logout of a Pod, etc.
     *
     * This term provides descriptions only in English.
     */
    Service: _NS$2("Service"),
    /**
     * An instance of solid_sd:Feature represents a feature of a
   Solid service. Specific types of features might include protocols (or
   sub-protocols)
     *
     * The term has a description only in English, with [1] label in the language [NoLocale], but [1] comment in the language [en] (so the difference is only between English and NoLocale, which we consider the same).
     */
    Feature: _NS$2("Feature"),
    /**
     * An instance of solid_sd:Protocol represents one specific
   protocol (or sub-protocol).
     *
     * The term has a description only in English, with [1] label in the language [NoLocale], but [1] comment in the language [en] (so the difference is only between English and NoLocale, which we consider the same).
     */
    Protocol: _NS$2("Protocol"),
    // *******************
    // All the Properties.
    // *******************
    /**
     * A Solid Endpoint is the URL through which an agent requests
  that a Solid Service performs its function, e.g., the SPARQL endpoint for a
  "Search my Pod Solid Service.
     *
     * This term provides descriptions only in English.
     */
    endpoint: _NS$2("endpoint"),
    /**
     * Relates an instance of solid_sd:Service with a
   resource representing a supported feature.
     *
     * The term has a description only in English, with [1] label in the language [NoLocale], but [1] comment in the language [en] (so the difference is only between English and NoLocale, which we consider the same).
     */
    feature: _NS$2("feature"),
    /**
     * Relates an instance of solid_sd:Service to a protocol (or
   sub-protocol) that it implements (e.g., a WebSocket service might implement
   multiple, or just one, sub-protocols.
     *
     * This term provides descriptions only in English.
     */
    supportedProtocol: _NS$2("supportedProtocol"),
    /**
     * A Solid Service providing SPARQL query capability - could be
   across data within a single Pod, or across all Pods hosted by a Solid Server
   provider.
     *
     * This term provides descriptions only in English.
     */
    serviceSparql: _NS$2("serviceSparql"),
    /**
     * A Solid Service providing GraphQL query capability. This
   service initially provides authenticated and authorized access to sensitive
   access control data associated with resources in Pods.
     *
     * This term provides descriptions only in English.
     */
    serviceGraphql: _NS$2("serviceGraphql"),
    /**
     * A Solid Service providing the ability to 'power off' a Pod.
   This is useful in cases where a Pod owner wishes to completely disable all
   access to their entire Pod (e.g., in the case that they suspect access has been
   compromised).
     *
     * This term provides descriptions only in English.
     */
    servicePowerSwitch: _NS$2("servicePowerSwitch"),
    /**
     * A Solid Service providing the ability to register a new
   a Pod. Generally wed expect this service to be provided by a Solid
   Identity Provider, although some server implementations may wish to offer this service directly themselves too.
     *
     * This term provides descriptions only in English.
     */
    servicePodRegistration: _NS$2("servicePodRegistration"),
    /**
     * A Solid Service providing the ability to subscribe to
   resource change events via a WebSocket connection.
     *
     * This term provides descriptions only in English.
     */
    serviceWebSocket: _NS$2("serviceWebSocket"),
    /**
     * solid-0.2
     *
     * This term has a label (in language [NoLocale]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
     *
     * See also:
     *  - http://www.w3.org/ns/solid/service-description#WebSocketSubProtocol0_2
     */
    WebSocketSubProtocol0_2Label: _NS$2("WebSocketSubProtocol0_2Label"),
    /**
     * solid-0.1
     *
     * This term has a label (in language [NoLocale]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
     *
     * See also:
     *  - http://www.w3.org/ns/solid/service-description#WebSocketSubProtocol0_1
     */
    WebSocketSubProtocol0_1Label: _NS$2("WebSocketSubProtocol0_1Label"),
};

/**
 * MIT License
 *
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * The Solid vocabulary providing terms (e.g., classes, properties and text strings (e.g.
 informative labels or error messages) commonly used across Solid services.
 */
// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
var _NAMESPACE$1 = "https://inrupt.com/vocab/solid_service_core#";
function _NS$1(localName) {
    return (_NAMESPACE$1 + localName);
}
/**
 * The Solid vocabulary providing terms (e.g., classes, properties and text strings (e.g.
 informative labels or error messages) commonly used across Solid services.
 */
var SOLID_SERVICE_CORE = {
    PREFIX: "solid_service_core",
    NAMESPACE: _NAMESPACE$1,
    PREFIX_AND_NAMESPACE: { "solid_service_core": "https://inrupt.com/vocab/solid_service_core#" },
    NS: _NS$1,
    // *******************
    // All the Literals.
    // *******************
    /**
     * Failed to process incoming message as RDF
     *
     * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
     *
     * Defined by the vocabulary: https://inrupt.com/vocab/solid_service_core#
     */
    errFailedToProcessIncomingRdf: _NS$1("errFailedToProcessIncomingRdf"),
    /**
     * Incoming request validated as RDF (it contains [{{0}}] triples!)
     *
     * This term has no descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:label&#x27;, &#x27;rdfs:comment&#x27;, or &#x27;dcterms:description&#x27; meta-data).
     *
     * Defined by the vocabulary: https://inrupt.com/vocab/solid_service_core#
     */
    msgRequestValidatedAsRdf: _NS$1("msgRequestValidatedAsRdf"),
};

/**
 * MIT License
 *
 * Copyright 2022 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.4]
 * as part of artifact: [vocab-solid], version: [1.0.4]
 * on 'Wednesday, June 1, 2022 7:54 PM'.
 *
 * Vocabulary built from vocab list file: [solid-rdf/vocab-solid.yml].
 *
 * This extension adds predicates making explicit some implicit knowledge in the Workspace ontology.
 */
// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
var _NAMESPACE = "http://www.w3.org/ns/pim/space#";
function _NS(localName) {
    return (_NAMESPACE + localName);
}
/**
 * This extension adds predicates making explicit some implicit knowledge in the Workspace ontology.
 */
var WS = {
    PREFIX: "ws",
    NAMESPACE: _NAMESPACE,
    PREFIX_AND_NAMESPACE: { "ws": "http://www.w3.org/ns/pim/space#" },
    NS: _NS,
    // *****************
    // All the Classes.
    // *****************
    /**
     * A file describing a configuration. In this context, it is
   expected to describe the workspace configuration.
     *
     * This term provides multilingual descriptions, but has a mismatch between its labels and comments, with [2] labels in languages [en, fr], but [1] comment in the language [en].
     */
    ConfigurationFile: _NS("ConfigurationFile"),
    /**
     * A  storage is a space of URIs in which you can individually control for each resource
      who has access to it.
  
     *
     * This term provides descriptions only with no explicit locale.
     */
    ControlledStorage: _NS("ControlledStorage"),
    /**
     * A storage is a space of URIs in which you have access to data.
  
     *
     * This term provides descriptions only with no explicit locale.
     */
    Storage: _NS("Storage"),
    /**
     * This is a workspace for storing the
      information about the other workspaces.
      As a user, you normally don't have to worry about it.
     *
     * The term has a description only in English, with [1] label in the language [en], but [1] comment in the language [NoLocale] (so the difference is only between English and NoLocale, which we consider the same).
     */
    MasterWorkspace: _NS("MasterWorkspace"),
    /**
     * Access only by the you, the user.
     *
     * This term provides descriptions only with no explicit locale.
     */
    PrivateWorkspace: _NS("PrivateWorkspace"),
    /**
     * Workspaces are place where data is stored, and associated polices of privacy.
  A given application typically stores information in several different
  workspaces, some being user private, some shared, and some public.
  
     *
     * This term provides descriptions only with no explicit locale.
     */
    Workspace: _NS("Workspace"),
    /**
     * A personal storage is a space of URIs in which you and only you have access to data,
      you cannot give access to anyone else.
  
     *
     * This term provides descriptions only with no explicit locale.
     */
    PersonalStorage: _NS("PersonalStorage"),
    /**
     * Aceess may not be open to the public. Contains preferences resources.
     *
     * This term provides descriptions only with no explicit locale.
     */
    PreferencesWorkspace: _NS("PreferencesWorkspace"),
    /**
     * Access is to some but not all people.
     *
     * This term provides descriptions only with no explicit locale.
     */
    SharedWorkspace: _NS("SharedWorkspace"),
    /**
     * A public storage is a space of URIs in which you have access to data,
      and all data is accessible to anyone without control.
  
     *
     * This term provides descriptions only with no explicit locale.
     */
    PublicStorage: _NS("PublicStorage"),
    /**
     * Aceess is open to the public. Anything in a public workspace
      can be accesed by anyone.
     *
     * This term provides descriptions only with no explicit locale.
     */
    PublicWorkspace: _NS("PublicWorkspace"),
    // *******************
    // All the Properties.
    // *******************
    /**
     * master workspace
     *
     * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
     */
    masterWorkspace: _NS("masterWorkspace"),
    /**
     * preferences file
     *
     * This term has a label (in language [NoLocale]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
     */
    preferencesFile: _NS("preferencesFile"),
    /**
     * The storage in which this workspace is, or the storage which
  contains this resource, or a storage available to this agent to use.
     *
     * This term provides descriptions only with no explicit locale.
     */
    storage: _NS("storage"),
    /**
     * URIs which start with this string are in this workspace or storage.
  This may be used for constructing URIs for new storage resources.
  
     *
     * This term provides descriptions only with no explicit locale.
     */
    uriPrefix: _NS("uriPrefix"),
    /**
     * workspace
     *
     * This term has a label (in language [en]), but no long-form descriptions at all (i.e., the vocabulary doesn&#x27;t provide any &#x27;rdfs:comment&#x27; or &#x27;dcterms:description&#x27; meta-data).
     */
    workspace: _NS("workspace"),
};

exports.ACL = ACL;
exports.ACP = ACP;
exports.OIDC = OIDC;
exports.SOLID = SOLID;
exports.SOLID_META_ACL = SOLID_META_ACL;
exports.SOLID_META_CLIENT = SOLID_META_CLIENT;
exports.SOLID_META_SERVER_CREATIONAL = SOLID_META_SERVER_CREATIONAL;
exports.SOLID_META_SERVER_RUNTIME = SOLID_META_SERVER_RUNTIME;
exports.SOLID_SD = SOLID_SD;
exports.SOLID_SERVICE_CORE = SOLID_SERVICE_CORE;
exports.WS = WS;
