// @inrupt/artifact-generator/test/Generated/UNIT_TEST/SupportedDataType/owl-test/Generated/SourceCodeArtifacts/JavaScript/GeneratedVocab/OWL.js
/**
 * Generated by the Inrupt Artifact Generator [TestGenerator], version [^7.8.9]
 * as part of artifact: [generated-vocab-owl], version: [1.0.0]
 * on '1999/12/31 23:59'.
 *
 * Vocabulary built from input: [./test/resources/vocab/special-case-owl-snippet.ttl].
 * The generator detected the following terms in the source vocabulary:
 *  - Classes: [1]
 *  - Properties: [0]
 *  - Literals: [0]
 *  - Constant IRIs: [0]
 *  - Constant strings: [0]
 *
 * 
  This ontology partially describes the built-in classes and
  properties that together form the basis of the RDF/XML syntax of OWL 2.
  The content of this ontology is based on Tables 6.1 and 6.2
  in Section 6.4 of the OWL 2 RDF-Based Semantics specification,
  available at http://www.w3.org/TR/owl2-rdf-based-semantics/.
  Please note that those tables do not include the different annotations
  (labels, comments and rdfs:isDefinedBy links) used in this file.
  Also note that the descriptions provided in this ontology do not
  provide a complete and correct formal description of either the syntax
  or the semantics of the introduced terms (please see the OWL 2
  recommendations for the complete and normative specifications).
  Furthermore, the information provided by this ontology may be
  misleading if not used with care. This ontology SHOULD NOT be imported
  into OWL ontologies. Importing this file into an OWL 2 DL ontology
  will cause it to become an OWL 2 Full ontology and may have other,
  unexpected, consequences.
   
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const { VocabTerm: _VocabTerm, getLocalStore } = require("@inrupt/solid-common-vocab");
const _DataFactory = require("@rdfjs/data-model");

const { namedNode: _namedNode } = _DataFactory;

function _NS(localName) {
  return _namedNode(`http://www.w3.org/2002/07/owl#${localName}`);
}

/**
 * 
  This ontology partially describes the built-in classes and
  properties that together form the basis of the RDF/XML syntax of OWL 2.
  The content of this ontology is based on Tables 6.1 and 6.2
  in Section 6.4 of the OWL 2 RDF-Based Semantics specification,
  available at http://www.w3.org/TR/owl2-rdf-based-semantics/.
  Please note that those tables do not include the different annotations
  (labels, comments and rdfs:isDefinedBy links) used in this file.
  Also note that the descriptions provided in this ontology do not
  provide a complete and correct formal description of either the syntax
  or the semantics of the introduced terms (please see the OWL 2
  recommendations for the complete and normative specifications).
  Furthermore, the information provided by this ontology may be
  misleading if not used with care. This ontology SHOULD NOT be imported
  into OWL ontologies. Importing this file into an OWL 2 DL ontology
  will cause it to become an OWL 2 Full ontology and may have other,
  unexpected, consequences.
   
 *
 * Namespace IRI: [http://www.w3.org/2002/07/owl#]
 */
const OWL = {
  PREFIX: "owl",
  NAMESPACE: "http://www.w3.org/2002/07/owl#",
  PREFIX_AND_NAMESPACE: { "owl": "http://www.w3.org/2002/07/owl#" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * The class of collections of pairwise different individuals.
   *
   * This term provides descriptions only with no explicit locale.
   *
   * Defined by the vocabulary: http://www.w3.org/2002/07/owl#
   */
  AllDifferent: new _VocabTerm(
    _NS("AllDifferent"),
    _DataFactory,
    getLocalStore(),
    false
  )
    .addType(_namedNode("http://www.w3.org/2000/01/rdf-schema#Class"))
    .addIsDefinedBy(_namedNode("http://www.w3.org/2002/07/owl#"))
    .addLabelNoLanguage(`AllDifferent`)
    .addCommentNoLanguage(`The class of collections of pairwise different individuals.`),

}

module.exports = OWL;
