// @inrupt/artifact-generator/test/Generated/UNIT_TEST/EndToEnd/isDefinedBy/Generated/SourceCodeArtifacts/JavaScript/GeneratedVocab/SCHEMA_INRUPT_EXT.js
/**
 * Generated by the Inrupt Artifact Generator [@inrupt/artifact-generator], version [3.1.1]
 * as part of artifact: [generated-vocab-common-TEST], version: [10.11.12]
 * on 'Monday, October 2, 2023 11:19 AM'.
 *
 * Vocabulary built from vocab list file: [./test/resources/yamlConfig/vocab-strict.yml].
 * The generator detected the following terms in the source vocabulary:
 *  - Classes: [1]
 *  - Properties: [2]
 *  - Literals: [0]
 *  - Constant IRIs: [0]
 *  - Constant strings: [0]
 *
 * Inrupt extension to Schema.org terms.
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const { VocabTerm: _VocabTerm, getLocalStore } = require("@inrupt/solid-common-vocab");
const _DataFactory = require("@rdfjs/data-model");

const { namedNode: _namedNode } = _DataFactory;

function _NS(localName) {
  return _namedNode(`https://schema.org/${localName}`);
}

/**
 * Inrupt extension to Schema.org terms.
 *
 * Namespace IRI: [https://schema.org/]
 */
const SCHEMA_INRUPT_EXT = {
  PREFIX: "schema_inrupt_ext",
  NAMESPACE: "https://schema.org/",
  PREFIX_AND_NAMESPACE: { "schema_inrupt_ext": "https://schema.org/" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * A person (alive, dead, undead, or fictional).
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   */
  Person: new _VocabTerm(
    _NS("Person"),
    _DataFactory,
    getLocalStore(),
    true
  )
    .addType(_namedNode("http://www.w3.org/2000/01/rdf-schema#Class"))
    .addLabelNoLanguage(`Person`)
    .addLabel(`Person`, "de")
    .addLabel(`Person`, "en")
    .addLabel(`Persona`, "es")
    .addLabel(`La personne`, "fr")
    .addLabel(`Persona`, "it")
    .addComment(`Une personne (vivante, morte, mort-vivant ou fictive).`, "fr")
    .addComment(`Eine Person (lebendig, tot, untot oder fiktiv).`, "de")
    .addComment(`Una persona (viva, muerta, no muerta o ficticia).`, "es")
    .addComment(`Una persona (viva, morta, non morta o immaginaria).`, "it")
    .addCommentNoLanguage(`A person (alive, dead, undead, or fictional).`),


  // *******************
  // All the Properties.
  // *******************

  /**
   * Given name. In the U.S., the first name of a Person. This can be used along with familyName instead of the name property.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   */
  givenName: new _VocabTerm(
    _NS("givenName"),
    _DataFactory,
    getLocalStore(),
    true
  )
    .addType(_namedNode("http://www.w3.org/1999/02/22-rdf-syntax-ns#Property"))
    .addLabelNoLanguage(`givenName`)
    .addLabel(`Vorname`, "de")
    .addLabel(`Given Name`, "en")
    .addLabel(`Nombre de pila`, "es")
    .addLabel(`Prénom`, "fr")
    .addLabel(`Nome di battesimo`, "it")
    .addComment(`Prénom. Aux États-Unis, le prénom d’une personne. Ceci peut être utilisé avec familyName au lieu de la propriété name.`, "fr")
    .addComment(`Vorname. In den USA der Vorname einer Person. Dies kann zusammen mit familyName anstelle der Eigenschaft name verwendet werden.`, "de")
    .addComment(`Nombre de pila. En los EE. UU., El primer nombre de una persona. Esto se puede usar junto con familyName en lugar de la propiedad name.`, "es")
    .addComment(`Nome di battesimo. Negli Stati Uniti, il primo nome di una persona. Questo può essere usato insieme a familyName al posto della proprietà name.`, "it")
    .addCommentNoLanguage(`Given name. In the U.S., the first name of a Person. This can be used along with familyName instead of the name property.`),

  /**
   * Family name. In the U.S., the last name of an Person. This can be used along with givenName instead of the name property.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   */
  familyName: new _VocabTerm(
    _NS("familyName"),
    _DataFactory,
    getLocalStore(),
    true
  )
    .addType(_namedNode("http://www.w3.org/1999/02/22-rdf-syntax-ns#Property"))
    .addLabelNoLanguage(`familyName`)
    .addLabel(`Nachname`, "de")
    .addLabel(`Family Name`, "en")
    .addLabel(`Apellido`, "es")
    .addLabel(`Nom de famille`, "fr")
    .addLabel(`Cognome`, "it")
    .addComment(`Nom de famille. Aux États-Unis, le nom de famille d’une personne. Ceci peut être utilisé avec GivenName au lieu de la propriété name.`, "fr")
    .addComment(`Nachname. In den USA der Nachname einer Person. Dies kann zusammen mit givenName anstelle der Eigenschaft name verwendet werden.`, "de")
    .addComment(`Apellido. En los EE.UU., el apellido de una persona. Esto se puede usar junto con givenName en lugar de la propiedad name.`, "es")
    .addComment(`Cognome. Negli Stati Uniti, il cognome di una persona. Questo può essere usato insieme a givenName al posto della proprietà name.`, "it")
    .addCommentNoLanguage(`Family name. In the U.S., the last name of an Person. This can be used along with givenName instead of the name property.`),

}

module.exports = SCHEMA_INRUPT_EXT;
